<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lecture 7 Ecological assembly | Theoretical Community Ecology</title>
  <meta name="description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Lecture 7 Ecological assembly | Theoretical Community Ecology" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lecture 7 Ecological assembly | Theoretical Community Ecology" />
  
  <meta name="twitter:description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  

<meta name="author" content="Stefano Allesina" />


<meta name="date" content="2025-12-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="game-theory-and-replicator-dynamics.html"/>
<link rel="next" href="metapopulation-dynamics-on-networks.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Theoretical Community Ecology</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preliminaries</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#approach-and-target-audience"><i class="fa fa-check"></i>Approach and target audience</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#topics"><i class="fa fa-check"></i>Topics</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#notation"><i class="fa fa-check"></i>Notation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#grading-and-code-of-conduct"><i class="fa fa-check"></i>Grading and code of conduct</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#homework-80"><i class="fa fa-check"></i>Homework (80%)</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#review-20"><i class="fa fa-check"></i>Review (20%)</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#accommodation-for-students-with-disabilities"><i class="fa fa-check"></i>Accommodation for students with disabilities</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#a-note-about-biographies"><i class="fa fa-check"></i>A note about biographies</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#sources"><i class="fa fa-check"></i>Sources</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html"><i class="fa fa-check"></i><b>1</b> Models for a single population</a>
<ul>
<li class="chapter" data-level="1.1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#types-of-dynamical-systems"><i class="fa fa-check"></i><b>1.1</b> Types of dynamical systems</a></li>
<li class="chapter" data-level="1.2" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#initial-value-problems"><i class="fa fa-check"></i><b>1.2</b> Initial-value problems</a></li>
<li class="chapter" data-level="1.3" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#solvable-model-exponential-growth"><i class="fa fa-check"></i><b>1.3</b> Solvable model: Exponential growth</a></li>
<li class="chapter" data-level="1.4" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#solvable-model-logistic-growth"><i class="fa fa-check"></i><b>1.4</b> Solvable model: Logistic growth</a></li>
<li class="chapter" data-level="1.5" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#qualitative-analysis-of-models-for-a-single-population"><i class="fa fa-check"></i><b>1.5</b> Qualitative analysis of models for a single population</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#stability-using-derivatives"><i class="fa fa-check"></i><b>1.5.1</b> Stability using derivatives</a></li>
<li class="chapter" data-level="1.5.2" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#what-are-we-doing-when-probing-stability-using-derivatives"><i class="fa fa-check"></i><b>1.5.2</b> What are we doing when probing stability using derivatives?</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#bifurcations"><i class="fa fa-check"></i><b>1.6</b> Bifurcations</a></li>
<li class="chapter" data-level="1.7" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#long-term-behavior-of-1-d-systems"><i class="fa fa-check"></i><b>1.7</b> Long term behavior of 1-D systems</a></li>
<li class="chapter" data-level="1.8" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#lyapunov-functions"><i class="fa fa-check"></i><b>1.8</b> Lyapunov functions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html"><i class="fa fa-check"></i><b>2</b> Models for two populations</a>
<ul>
<li class="chapter" data-level="2.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#qualitative-analysis-of-models-for-two-populations"><i class="fa fa-check"></i><b>2.1</b> Qualitative analysis of models for two populations</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#isoclines-of-null-growth"><i class="fa fa-check"></i><b>2.1.1</b> Isoclines of null growth</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#local-stability-analysis"><i class="fa fa-check"></i><b>2.2</b> Local stability analysis</a></li>
<li class="chapter" data-level="2.3" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#stability-analysis-of-the-lotka-volterra-predator-prey-model"><i class="fa fa-check"></i><b>2.3</b> Stability analysis of the Lotka-Volterra Predator-Prey model</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#constant-of-motion-for-lotka-volterra-predator-prey"><i class="fa fa-check"></i><b>2.3.1</b> Constant of motion for Lotka-Volterra Predator-Prey</a></li>
<li class="chapter" data-level="2.3.2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#analysis-of-predator-prey-model-with-logistic-growth-for-the-prey"><i class="fa fa-check"></i><b>2.3.2</b> Analysis of Predator-Prey model with logistic growth for the prey</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html"><i class="fa fa-check"></i><b>3</b> Generalized Lotka-Volterra model</a>
<ul>
<li class="chapter" data-level="3.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#formulation"><i class="fa fa-check"></i><b>3.1</b> Formulation</a></li>
<li class="chapter" data-level="3.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#a-single-population"><i class="fa fa-check"></i><b>3.2</b> A single population</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#metapopulation-dynamics"><i class="fa fa-check"></i><b>3.2.1</b> Metapopulation dynamics</a></li>
<li class="chapter" data-level="3.2.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#s-i-s-model"><i class="fa fa-check"></i><b>3.2.2</b> S-I-S model</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#multi-species-dynamics"><i class="fa fa-check"></i><b>3.3</b> Multi-species dynamics</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#existence-of-an-equilibrium"><i class="fa fa-check"></i><b>3.3.1</b> Existence of an equilibrium</a></li>
<li class="chapter" data-level="3.3.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#types-of-dynamics"><i class="fa fa-check"></i><b>3.3.2</b> Types of dynamics</a></li>
<li class="chapter" data-level="3.3.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#equilibrium-as-time-average"><i class="fa fa-check"></i><b>3.3.3</b> Equilibrium as time-average</a></li>
<li class="chapter" data-level="3.3.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#coexistence-requires-a-feasible-equilibrium"><i class="fa fa-check"></i><b>3.3.4</b> Coexistence requires a feasible equilibrium</a></li>
<li class="chapter" data-level="3.3.5" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#local-stability-of-the-coexistence-equilibrium"><i class="fa fa-check"></i><b>3.3.5</b> Local stability of the coexistence equilibrium</a></li>
<li class="chapter" data-level="3.3.6" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#lyapunov-diagonal-stability-and-global-stability"><i class="fa fa-check"></i><b>3.3.6</b> Lyapunov diagonal stability and global stability</a></li>
<li class="chapter" data-level="3.3.7" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#saturated-equilibria"><i class="fa fa-check"></i><b>3.3.7</b> Saturated equilibria</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#macarthurs-consumer-resource-model"><i class="fa fa-check"></i><b>3.4</b> MacArthur’s consumer-resource model</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#existence-of-an-equilibrium-1"><i class="fa fa-check"></i><b>3.4.1</b> Existence of an equilibrium</a></li>
<li class="chapter" data-level="3.4.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#global-stability-1"><i class="fa fa-check"></i><b>3.4.2</b> Global stability</a></li>
<li class="chapter" data-level="3.4.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#separation-of-time-scales"><i class="fa fa-check"></i><b>3.4.3</b> Separation of time-scales</a></li>
<li class="chapter" data-level="3.4.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#lyapunov-function-for-symmetric-lotka-volterra"><i class="fa fa-check"></i><b>3.4.4</b> Lyapunov function for symmetric Lotka-Volterra</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#further-readings"><i class="fa fa-check"></i><b>3.5</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html"><i class="fa fa-check"></i><b>4</b> Stability of large ecological communities</a>
<ul>
<li class="chapter" data-level="4.1" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#complexity-and-stability"><i class="fa fa-check"></i><b>4.1</b> Complexity and stability</a></li>
<li class="chapter" data-level="4.2" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#the-stability-of-random-ecosystems"><i class="fa fa-check"></i><b>4.2</b> The stability of random ecosystems</a></li>
<li class="chapter" data-level="4.3" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#accounting-for-interaction-types"><i class="fa fa-check"></i><b>4.3</b> Accounting for interaction types</a></li>
<li class="chapter" data-level="4.4" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#symmetric-matrices"><i class="fa fa-check"></i><b>4.4</b> Symmetric matrices</a></li>
<li class="chapter" data-level="4.5" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#covariance-matrices"><i class="fa fa-check"></i><b>4.5</b> Covariance matrices</a></li>
<li class="chapter" data-level="4.6" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#small-rank-perturbations"><i class="fa fa-check"></i><b>4.6</b> Small-rank perturbations</a></li>
<li class="chapter" data-level="4.7" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#structured-matrices"><i class="fa fa-check"></i><b>4.7</b> Structured matrices</a></li>
<li class="chapter" data-level="4.8" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#further-readings-1"><i class="fa fa-check"></i><b>4.8</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html"><i class="fa fa-check"></i><b>5</b> Predicting coexistence in ecological communities</a>
<ul>
<li class="chapter" data-level="5.1" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#background-diversity-and-ecosystem-functioning"><i class="fa fa-check"></i><b>5.1</b> Background: Diversity and Ecosystem Functioning</a></li>
<li class="chapter" data-level="5.2" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#example-data"><i class="fa fa-check"></i><b>5.2</b> Example data</a></li>
<li class="chapter" data-level="5.3" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#a-simple-statistical-framework"><i class="fa fa-check"></i><b>5.3</b> A simple statistical framework</a></li>
<li class="chapter" data-level="5.4" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#the-structure-of-equilibria-in-glv"><i class="fa fa-check"></i><b>5.4</b> The structure of equilibria in GLV</a></li>
<li class="chapter" data-level="5.5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#predicting-coexistence"><i class="fa fa-check"></i><b>5.5</b> Predicting coexistence</a></li>
<li class="chapter" data-level="5.6" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#scoring-matrices-b"><i class="fa fa-check"></i><b>5.6</b> Scoring matrices <span class="math inline">\(B\)</span></a></li>
<li class="chapter" data-level="5.7" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#example"><i class="fa fa-check"></i><b>5.7</b> Example</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#sum-of-squares"><i class="fa fa-check"></i><b>5.7.1</b> Sum of squares</a></li>
<li class="chapter" data-level="5.7.2" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#weighted-leas-squares"><i class="fa fa-check"></i><b>5.7.2</b> Weighted leas squares</a></li>
<li class="chapter" data-level="5.7.3" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#maximum-likelihood"><i class="fa fa-check"></i><b>5.7.3</b> Maximum likelihood</a></li>
<li class="chapter" data-level="5.7.4" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#leave-one-out-cross-validation"><i class="fa fa-check"></i><b>5.7.4</b> Leave-one-out cross validation</a></li>
<li class="chapter" data-level="5.7.5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#simplified-models"><i class="fa fa-check"></i><b>5.7.5</b> Simplified models</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#further-readings-2"><i class="fa fa-check"></i><b>5.8</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html"><i class="fa fa-check"></i><b>6</b> Game theory and replicator dynamics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#game-theory"><i class="fa fa-check"></i><b>6.1</b> Game theory</a></li>
<li class="chapter" data-level="6.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#bimatrix-games"><i class="fa fa-check"></i><b>6.2</b> Bimatrix games</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#example-the-prisoners-dilemma"><i class="fa fa-check"></i><b>6.2.1</b> Example: the prisoner’s dilemma</a></li>
<li class="chapter" data-level="6.2.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#mixed-strategies"><i class="fa fa-check"></i><b>6.2.2</b> Mixed strategies</a></li>
<li class="chapter" data-level="6.2.3" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#nash-equilibria"><i class="fa fa-check"></i><b>6.2.3</b> Nash Equilibria</a></li>
<li class="chapter" data-level="6.2.4" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#symmetric-bimatrix-games"><i class="fa fa-check"></i><b>6.2.4</b> Symmetric bimatrix games</a></li>
<li class="chapter" data-level="6.2.5" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#evolutionary-stable-strategies"><i class="fa fa-check"></i><b>6.2.5</b> Evolutionary stable strategies</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#replicator-dynamics"><i class="fa fa-check"></i><b>6.3</b> Replicator dynamics</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#nash-equilibria-are-equilibria-of-the-re"><i class="fa fa-check"></i><b>6.3.1</b> Nash equilibria are equilibria of the RE</a></li>
<li class="chapter" data-level="6.3.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#ess-are-globally-stable-equilibria-of-the-re"><i class="fa fa-check"></i><b>6.3.2</b> ESS are globally stable equilibria of the RE</a></li>
<li class="chapter" data-level="6.3.3" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#stability-of-the-re-with-symmetric-a"><i class="fa fa-check"></i><b>6.3.3</b> Stability of the RE with symmetric <span class="math inline">\(A\)</span></a></li>
<li class="chapter" data-level="6.3.4" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#invariants"><i class="fa fa-check"></i><b>6.3.4</b> Invariants</a></li>
<li class="chapter" data-level="6.3.5" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#rock-paper-scissor"><i class="fa fa-check"></i><b>6.3.5</b> Rock-paper-scissor</a></li>
<li class="chapter" data-level="6.3.6" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#equivalence-with-glv"><i class="fa fa-check"></i><b>6.3.6</b> Equivalence with GLV</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#hypertournament-games"><i class="fa fa-check"></i><b>6.4</b> Hypertournament games</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#lyapunov-function"><i class="fa fa-check"></i><b>6.4.1</b> Lyapunov function</a></li>
<li class="chapter" data-level="6.4.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#higher-order-interactions"><i class="fa fa-check"></i><b>6.4.2</b> Higher-order interactions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ecological-assembly.html"><a href="ecological-assembly.html"><i class="fa fa-check"></i><b>7</b> Ecological assembly</a>
<ul>
<li class="chapter" data-level="7.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#assembly"><i class="fa fa-check"></i><b>7.1</b> Assembly</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#what-makes-the-study-of-assembly-difficult"><i class="fa fa-check"></i><b>7.1.1</b> What makes the study of assembly difficult?</a></li>
<li class="chapter" data-level="7.1.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#ecological-assembly-without-tears"><i class="fa fa-check"></i><b>7.1.2</b> Ecological assembly without tears</a></li>
<li class="chapter" data-level="7.1.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#top-down-and-bottom-up-assembly"><i class="fa fa-check"></i><b>7.1.3</b> Top-down and bottom-up assembly</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#modeling-invasions"><i class="fa fa-check"></i><b>7.2</b> Modeling invasions</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#invasions-in-multispecies-communities"><i class="fa fa-check"></i><b>7.2.1</b> Invasions in multispecies communities</a></li>
<li class="chapter" data-level="7.2.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#lyapunov-stability-and-saturated-equilibria"><i class="fa fa-check"></i><b>7.2.2</b> Lyapunov stability and saturated equilibria</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#top-down-assembly"><i class="fa fa-check"></i><b>7.3</b> Top-down assembly</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#a-trivial-case"><i class="fa fa-check"></i><b>7.3.1</b> A trivial case</a></li>
<li class="chapter" data-level="7.3.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#the-random-zoo"><i class="fa fa-check"></i><b>7.3.2</b> The random zoo</a></li>
<li class="chapter" data-level="7.3.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#simulating-final-composition"><i class="fa fa-check"></i><b>7.3.3</b> Simulating final composition</a></li>
<li class="chapter" data-level="7.3.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#a-random-zoo-in-the-wild"><i class="fa fa-check"></i><b>7.3.4</b> A random zoo in the wild</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#bottom-up-assembly"><i class="fa fa-check"></i><b>7.4</b> Bottom-up assembly</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#simulating-bottom-up-assembly"><i class="fa fa-check"></i><b>7.4.1</b> Simulating bottom-up assembly</a></li>
<li class="chapter" data-level="7.4.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#network-spandrels"><i class="fa fa-check"></i><b>7.4.2</b> Network spandrels</a></li>
<li class="chapter" data-level="7.4.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#an-assembly-graph"><i class="fa fa-check"></i><b>7.4.3</b> An assembly graph</a></li>
<li class="chapter" data-level="7.4.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#how-many-invasions"><i class="fa fa-check"></i><b>7.4.4</b> How many invasions?</a></li>
<li class="chapter" data-level="7.4.5" data-path="ecological-assembly.html"><a href="ecological-assembly.html#properties-of-the-assembly-graph"><i class="fa fa-check"></i><b>7.4.5</b> Properties of the assembly graph</a></li>
<li class="chapter" data-level="7.4.6" data-path="ecological-assembly.html"><a href="ecological-assembly.html#assembly-graphs-for-glv"><i class="fa fa-check"></i><b>7.4.6</b> Assembly graphs for GLV</a></li>
<li class="chapter" data-level="7.4.7" data-path="ecological-assembly.html"><a href="ecological-assembly.html#relationship-with-lyapunov-functions"><i class="fa fa-check"></i><b>7.4.7</b> Relationship with Lyapunov functions</a></li>
<li class="chapter" data-level="7.4.8" data-path="ecological-assembly.html"><a href="ecological-assembly.html#how-many-assembly-endpoints"><i class="fa fa-check"></i><b>7.4.8</b> How many assembly endpoints?</a></li>
<li class="chapter" data-level="7.4.9" data-path="ecological-assembly.html"><a href="ecological-assembly.html#build-your-own-assembly-graph"><i class="fa fa-check"></i><b>7.4.9</b> Build your own assembly graph!</a></li>
<li class="chapter" data-level="7.4.10" data-path="ecological-assembly.html"><a href="ecological-assembly.html#conclusions"><i class="fa fa-check"></i><b>7.4.10</b> Conclusions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html"><i class="fa fa-check"></i><b>8</b> Metapopulation dynamics on networks</a>
<ul>
<li class="chapter" data-level="8.1" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#metapopulations"><i class="fa fa-check"></i><b>8.1</b> Metapopulations</a></li>
<li class="chapter" data-level="8.2" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#hanski-ovaskainen-2000"><i class="fa fa-check"></i><b>8.2</b> Hanski Ovaskainen 2000</a></li>
<li class="chapter" data-level="8.3" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#sis-on-a-contact-network"><i class="fa fa-check"></i><b>8.3</b> SIS on a contact network</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#the-full-model"><i class="fa fa-check"></i><b>8.3.1</b> The full model</a></li>
<li class="chapter" data-level="8.3.2" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#the-mean-field-approximation"><i class="fa fa-check"></i><b>8.3.2</b> The mean-field approximation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#metapopulations-on-randomly-distributed-patches"><i class="fa fa-check"></i><b>8.4</b> Metapopulations on randomly distributed patches</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html"><i class="fa fa-check"></i><b>9</b> Quasi-polynomial systems</a>
<ul>
<li class="chapter" data-level="9.1" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#global-stability-2"><i class="fa fa-check"></i><b>9.1</b> Global stability</a></li>
<li class="chapter" data-level="9.2" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#from-qp-to-glv"><i class="fa fa-check"></i><b>9.2</b> From QP to GLV</a></li>
<li class="chapter" data-level="9.3" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#stability-of-glv"><i class="fa fa-check"></i><b>9.3</b> Stability of GLV</a></li>
<li class="chapter" data-level="9.4" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#stability-for-glv-embeddings"><i class="fa fa-check"></i><b>9.4</b> Stability for GLV embeddings</a></li>
<li class="chapter" data-level="9.5" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#numerical-analysis"><i class="fa fa-check"></i><b>9.5</b> Numerical analysis</a></li>
<li class="chapter" data-level="9.6" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#systems-with-many-equations"><i class="fa fa-check"></i><b>9.6</b> Systems with many equations</a></li>
<li class="chapter" data-level="9.7" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#glv-embedding-for-non-qp-systems"><i class="fa fa-check"></i><b>9.7</b> GLV embedding for non-QP systems</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Theoretical Community Ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="ecological-assembly" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Lecture 7</span> Ecological assembly<a href="ecological-assembly.html#ecological-assembly" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><strong>Lesson plan</strong>:</p>
<ol style="list-style-type: decimal">
<li>We have seen in the previous lectures that a large ecological community will almost invariably be unstable, possibly leading to extinctions. It is therefore natural to ask what happens when either a feasible equilibrium does not exist, or it is unstable. In particular, we want to know how many species can persist.</li>
<li>We discuss the problem of assembly, and why is it so challenging to study analytically.</li>
<li>We introduce a few assumptions that make the problem tractable and at the same time non trivial.</li>
<li>We compute how many species will coexist in a GLV with random parameters, and see how we can study assembly by drawing an “assembly graph”</li>
</ol>
<div id="assembly" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Assembly<a href="ecological-assembly.html#assembly" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We call <strong>ecological assembly</strong> the process by which <strong>ecological communities are built</strong> by the interplay of invasions (increasing community “richness”) and extinctions (decreasing it). The typical setting for an assembly model is one in which a large (possibly infinite) pool of species are available (e.g., in a “mainland”), and every so often individuals from the species’ pool enter a local habitat (island, patch, etc.), at which point they might either grow (invasion, establishment, etc.), or go extinct. In either case, the local community might respond to the invader—for example, the invader could send some of the local populations extinct, or shift the community from one state to another.</p>
<p>Historically, this type of island-mainland models (based on species) have been contrasted with “trait-environment” models, in which species’ traits, rather than identity are the focus of the dynamics. In such models, ideas such as “environmental filtering” (i.e., the pruning of species that cannot grow in the local environment) and “trait underdispersion” (i.e., the fact that the filtering imposed by the environment results in a convergence in traits) arise naturally. When thinking of communities of competitors (e.g., different species of plants), then we might think that species’ interactions will also play a role—because species need to be different enough to coexist, this would generate “trait overdispersion” (due to limiting similarity). For a very opinionated summary of 25 years of discussions on these themes (1975-2000), see <span class="citation">Weiher and Keddy (<a href="#ref-weiher2001ecological">2001</a>)</span> (but first read the review <span class="citation">Gotelli (<a href="#ref-gotelli1999communities">1999</a>)</span>, stating that the language and tone of the introduction is “an embarrassment to the discipline”).</p>
<p>The idea of assembly was first introduced in ecology by the pioneers of “succession”. For example, in 1899 Henry Chandler Cowles studied the vegetation development in the Indiana Dunes—sand dunes are first colonized by “pioneer” species of plants, which are then overgrown by different vegetation in older dunes. Because dunes are regularly disrupted by wind, one could make a “space-for-time” substitution and reconstruct the development of the vegetation as a “chronosequence”. Importantly, Cowles (and then Clements) held the view that succession would be a strongly ordered, “deterministic” process by which communities tended to a certain “climax” state. In the 1920s, Henry Gleason challenged this view, advocating for a much greater role for chance.</p>
<div class="boxhistory">
<p><strong>History: Henry Chandler Cowles (1869-1939)</strong></p>
<p><img src="img/Cowles.jpg" width="35%" style="display: block; margin: auto;" /></p>
<p>Born in Kensington, Connecticut, Cowles attended Oberlin College, and received a PhD from the University of Chicago in 1898. His thesis is centered on the vegetation succession of the Indiana Dunes. His main inspiration came from the work of Eugen Warming (Cowles studied Danish in order to read the original <em>Plantesamfund</em>).</p>
<p>He was one of the pioneers of “ecology” in the US, a term that he helped popularizing, for example writing (along with Coulter and Barnes, all at U. Chicago) <em>A textbook of botany for colleges and universities</em>, with the second volume dedicated to the “Ecology” of plants. He was one of the founding members of the Ecological Society of America (in 1915).</p>
</div>
<p>Another important piece of the puzzle was provided by the work of Jared Diamond who, while studying avian assemblages in New Guinea, proposed that certain species assemblages would be “forbidden” (due to competition)—by observing several instances of assembly (in the different islands), one could guess the rules of the assembly game (<span class="citation">Diamond (<a href="#ref-diamond1975assembly">1975</a>)</span>). Diamond’s work sparked decades of intense debate, bringing a focus on the role of null models in ecology (are the observed patterns due to chance or necessity?), with especially cogent critiques brought forward by Simberloff, Connor and Gotelli (see for example <span class="citation">Gotelli and Graves (<a href="#ref-gotelli1996null">1996</a>)</span>).</p>
<div class="boxhistory">
<p><strong>History: Evelyn Chrystalla “E.C.” Pielou (1924-2016)</strong></p>
<p><img src="img/Chris_Pielou.png" width="35%" style="display: block; margin: auto;" /></p>
<p>Born in England, and developing her career in Canada, she was one of the first women in mathematical ecology. She studied radio-physics at the University of London, receiving her certificate at 18 years old. She went on to work in the Navy during WWII—meeting, and subsequently marrying a biologist. This fact changed the course of her career—having completed a Bachelor’s degree in Botany, she kept publishing research while raising three children and eventually earning a PhD (also from U. London). After a few years in the Canadian Dept. of Forestry (and then Agriculture), she was hired as a full professor at Queens U. (Ontario), and subsequently Dalhousie and Lethbrdige.</p>
<p>Her books “Introduction to Mathematical Ecology” and subsequent “Mathematical Ecology” (<span class="citation">Pielou (<a href="#ref-pielou1977mathematical">1977</a>)</span>)—which starts with: <em>“The fact that ecology is essentially a mathematical subject is becoming even more widely accepted”</em>—have formed generations of (theoretical) ecologists. As stated in her obituary <em>“Never afraid to express an opinion, she did not suffer fools gladly”</em>—as shown over and again in her work.</p>
<p>She is well-known for her books, for the development of null models in biogeography, and her measure of “evenness” of a community.</p>
</div>
<p>In the early 2000s, the focus changed decisively and became centered on the difference between “neutrality” (i.e., where species distribution are driven by stochastic fluctuations, <span class="citation">Hubbell (<a href="#ref-hubbell2001unified">2001</a>)</span>) and “niche” (i.e., driven by species interactions—or rather their avoidance).</p>
<p>Finally, the budding field of “community phylogenetics” (<span class="citation">Cadotte and Davies (<a href="#ref-cadotte2016phylogenies">2016</a>)</span>) is centered on relating these concepts to data stemming from phylogenetic trees.</p>
<p>Here, we step back and reconsider basic models of ecological assembly in the style for example of <span class="citation">Drake (<a href="#ref-drake1991community">1991</a>)</span>, <span class="citation">Law and Morton (<a href="#ref-law1996permanence">1996</a>)</span>, and <span class="citation">Morton et al. (<a href="#ref-morton1996models">1996</a>)</span>.</p>
<div id="what-makes-the-study-of-assembly-difficult" class="section level3 hasAnchor" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> What makes the study of assembly difficult?<a href="ecological-assembly.html#what-makes-the-study-of-assembly-difficult" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>There are three main complications that stand in the way of the development of simple theories for ecological assembly (taken from <span class="citation">Serván and Allesina (<a href="#ref-servan2020tractable">2020</a>)</span>):</p>
<ol style="list-style-type: decimal">
<li><p><strong>Invasion rates</strong>. We have a timescale for the local dynamics, and a timescale for invasions, and the time at which the first species goes extinct after an invasion influences the effect of subsequent invasions. The simplest way to think of this is to have a “rock-paper-scissors” community. Suppose that our island contains only “rock”, and that “paper” enters the system—if we wait for long enough, it will displace rock; if <em>before rock goes extinct</em>, “scissors” invades, we can recover the full three-species community; if instead scissors arrives when rock is extinct, then it will displace paper.</p>
<p>As such, if the speed at which the dynamics of the local community proceed are slow enough compared to the rate of invasion, we have that several species can invade before the community has reached its asymptotic configuration. At the extreme where local dynamics are fast compared to the rate of invasion, we have that each invader finds the local community at its asymptotic state; as the invasion rate increases, the system approaches a point where all the species enter the system before any extinction takes place. Increasing the invasion rate even further would result in an open system with constant immigration.</p></li>
<li><p><strong>Invasion size.</strong> Consider the two-species competitive Lotka-Volterra model with preemptive competition, and suppose that initially we have species <span class="math inline">\(x_1\)</span> resting at its carrying capacity, i.e., the state of the system is <span class="math inline">\(\{x_1\}\)</span>. If <span class="math inline">\(x_2\)</span> invades with sufficiently low density, we find <span class="math inline">\(\{x_1, x_2\} \to \{x_1\}\)</span>; on the other hand, if <span class="math inline">\(x_2\)</span> has sufficiently high density, we can cross the separatrix in the phase plane, leading to <span class="math inline">\(\{x_1, x_2\} \to \{x_2\}\)</span>. This simple example shows that the density at which the invader enters the system can alter the outcome of the dynamics.</p></li>
<li><p><strong>Invasion timing.</strong> When the local community coexists at a non-fixed point attractor, the fate of the invader could be very different depending on when it is introduced. For example, a predator requiring its prey to be above a certain level would not be able to invade an oscillating system whenever prey are at low abundance, but would start growing if the invasion happened at a time when prey were abundant.</p></li>
</ol>
</div>
<div id="ecological-assembly-without-tears" class="section level3 hasAnchor" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Ecological assembly without tears<a href="ecological-assembly.html#ecological-assembly-without-tears" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We make three main assumptions, which remove the difficulties above while leaving the model interesting to study:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Invasion events are rare.</strong> We assume that the invasion rate is low enough such that, after an invasion, the local community has sufficient time to reach its asymptotic configuration before the next invader arrives. In other words, we consider cases in which local dynamics operates at a much faster rate than invasions. Note that this choice precludes certain dynamics; for example, under these stringent conditions the rock-paper-scissors community described above would never reach the three-species configuration. While this is a strong requirement, it corresponds to assumptions routinely made in the study of population genetics, and in invasion analysis.</p></li>
<li><p><strong>Invaders arrive at low abundance.</strong> We assume that the density of the invader is low enough so that intraspecific effects are negligible at the time of invasion (again, as routinely assumed in invasion analysis). Under this assumption, the assembly of the Lotka-Volterra preemptive competition model would have two possible final states, corresponding to each species in isolation. Note also that whenever the invader can enter the system only at low abundance, the local stability of an attractor (i.e. the community at the attractor is resistant to small perturbations caused by changes in abundance of any of the species in the pool) is sufficient to make it a possible outcome of the assembly process.</p></li>
<li><p><strong>Fixed-point dynamics.</strong> Finally, we consider models in which the asymptotic state of the local community is a feasible, stable equilibrium, thereby sidestepping the difficulty stemming from the timing of invasion.</p>
<p>For example, the Generalized Lotka-Volterra model with a symmetric matrix of interaction <span class="math inline">\(A\)</span> yields only fixed-point dynamics.</p></li>
</ol>
</div>
<div id="top-down-and-bottom-up-assembly" class="section level3 hasAnchor" number="7.1.3">
<h3><span class="header-section-number">7.1.3</span> Top-down and bottom-up assembly<a href="ecological-assembly.html#top-down-and-bottom-up-assembly" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We distinguish between two extreme cases of assembly. In <strong>top-down</strong> assembly, all species enter the system (at arbitrary densities) <strong>at the same time</strong>; assembly then is simply the pruning of some of the species through the dynamics. At the other extreme, we find <strong>bottom-up assembly</strong>, in which species enter the system one at a time. Our second lecture will focus on top-down assembly, and the third on bottom-up assembly.</p>
</div>
</div>
<div id="modeling-invasions" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> Modeling invasions<a href="ecological-assembly.html#modeling-invasions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Imagine starting with a bare environment, and introducing the first species. If the species can grow in the new environment, it will establish, and if not, it will go extinct. Mathematically, we assume that <strong>new species are introduced at very low abundance</strong>, so that they do not experience self-limitation due to crowding, and that <strong>invasion are spaced in time</strong> so that the dynamics can play out before the next invasion happens. For example, consider the GLV model, and the case of the first species entering the system, and write the equation for the per-capita growth rate:</p>
<p><span class="math display">\[
\dfrac{1}{x_i(t)}\dfrac{dx_i(t)}{dt} = r_i + A_{ii} x_i(t)
\]</span></p>
<p>If <span class="math inline">\(x_i(0) \ll 1\)</span>, we can set <span class="math inline">\(A_{ii} x_i(0) \approx 0\)</span>, obtaining</p>
<p><span class="math display">\[
\dfrac{1}{x_i(t)}\dfrac{dx_i(t)}{dt} \approx r_i
\]</span></p>
<p>That is, the species will establish if it has a positive growth rate in the new environment. Suppose that this is the case: then species <span class="math inline">\(i\)</span> will grow to its equilibrium abundance <span class="math inline">\(x_i^\star = -r_i / A_{ii}\)</span>. Now add a second species. Its initial per-capita growth rate is going to be:</p>
<p><span class="math display">\[
\dfrac{1}{x_j(t)}\dfrac{dx_j(t)}{dt} = r_j + \sum_k A_{jk} x_k(t) \approx r_j + A_{ji}x_i^\star
\]</span></p>
<p>Species <span class="math inline">\(j\)</span> can therefore grow when rare if <span class="math inline">\(r_j + A_{ji}x_i^\star &gt; 0\)</span>, i.e., <span class="math inline">\(r_j &gt; -A_{ji}x_i^\star\)</span>. We call this type of inequality an <strong>“invasion criterion”</strong>.</p>
<p>If the species <span class="math inline">\(j\)</span> can grow when rare, in general, it could a) grow initially, but then go extinct; b) displace species <span class="math inline">\(i\)</span>, sending it to extinction; c) coexist with species <span class="math inline">\(i\)</span>. By reiterating invasions with different species, we can assemble a large ecological community.</p>
<div id="invasions-in-multispecies-communities" class="section level3 hasAnchor" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> Invasions in multispecies communities<a href="ecological-assembly.html#invasions-in-multispecies-communities" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now consider a pool of species (e.g., a metacommunity/mainland) and an environment (e.g., a local habitat/island) in which some of the species are present, and coexisting at an equilibrium. We have <span class="math inline">\(n\)</span> species in the pool, and <span class="math inline">\(k\)</span> species in the habitat/island. We want to write conditions for the invasibility of the equilibrium. To this end, we can re-arrange the rows and columns of <span class="math inline">\(A\)</span>, and the elements of <span class="math inline">\(r\)</span> to obtain two blocks: one for the <span class="math inline">\(k\)</span> species already in the community, and one for the <span class="math inline">\(n-k\)</span> potential invaders.</p>
<p>The fixed point <span class="math inline">\(\bar{x}\)</span> can be written as:</p>
<p><span class="math display">\[
\bar{x} = \left(\begin{array}{l}
x^{(k)} \\
0^{(n-k)}
\end{array} \right)
\]</span></p>
<p>where <span class="math inline">\(x^{(k)}\)</span> contains the density of the coexisting species. Similarly, the growth rates are</p>
<p><span class="math display">\[
r = \left(\begin{array}{l}
r^{(k)} \\
r^{(n-k)}
\end{array} \right)
\]</span></p>
<p>and the interaction matrix</p>
<p><span class="math display">\[
A = \begin{pmatrix}
A^{(k,k)} &amp; A^{(k, n- k)}\\
A^{(n-k,k)} &amp; A^{(n-k, n- k)}\\
  \end{pmatrix}
\]</span></p>
<p>where each block <span class="math inline">\(A^{(a,b)}\)</span> contains the effects of the species in set <span class="math inline">\(b\)</span> on the growth of the species in <span class="math inline">\(a\)</span>. For feasibility we need:</p>
<p><span class="math display">\[
x^{(k)} = -\left(A^{(k,k)} \right)^{-1}r^{(k)} &gt; 0
\]</span></p>
<p>Now we want to write the condition for the <strong>non-invasibility</strong> of the resident community by the other species in the metacommunity. For each species in <span class="math inline">\((n-k)\)</span>, we need to have a negative growth rate when invading:</p>
<p><span class="math display">\[
r^{(n-k)} + A^{(n-k,k)}x^{(k)} &lt; 0
\]</span></p>
<p>an equilibrium <span class="math inline">\(\bar{x}\)</span> for which a) the <span class="math inline">\(x^{(k)}\)</span> is feasible and stable (when considering only the species in <span class="math inline">\(k\)</span>) and b) no other species can invade when rare is called the <strong>feasible, stable, non-invasible solution</strong> or the <strong>saturated equilibrium</strong>.</p>
</div>
<div id="lyapunov-stability-and-saturated-equilibria" class="section level3 hasAnchor" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Lyapunov stability and saturated equilibria<a href="ecological-assembly.html#lyapunov-stability-and-saturated-equilibria" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We have seen in previous section that, whenever there exists a diagonal matrix with positive coefficients on the diagonal <span class="math inline">\(C\)</span> such that <span class="math inline">\(CA + A^T C\)</span> is negative definite, and we have a feasible equilibrium <span class="math inline">\(x^\star\)</span>, then the equilibrium is globally stable.</p>
<p>Now we consider the case in which a feasible equilibrium does not exist. However, we have proven that a <strong>saturated equilibrium</strong> exists and is unique: we have an equilibrium <span class="math inline">\(\bar{x}\)</span> in which some components are positive and some are zero such that <span class="math inline">\(k\)</span> species coexist at a globally stable equilibrium, and the remaining <span class="math inline">\((n-k)\)</span> species cannot invade it.</p>
</div>
</div>
<div id="top-down-assembly" class="section level2 hasAnchor" number="7.3">
<h2><span class="header-section-number">7.3</span> Top-down assembly<a href="ecological-assembly.html#top-down-assembly" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To begin our explorations of assembly, we consider the case in which all species are introduced in the habitat at the same time, and at arbitrary initial densities.</p>
<p>Here we investigate the simplest case in which dynamics are given by the GLV model, the (stable) matrix of interactions has random coefficients, and the growth rates are also random. For simplicity, we take the off-diagonal elements of <span class="math inline">\(A\)</span> and the growth rates from a distribution centered at zero. This case was studied by <span class="citation">Serván et al. (<a href="#ref-servan2018coexistence">2018</a>)</span>.</p>
<p>First, we are going to consider a trivial case, then consider a more complex one, and finally outline the proof in <span class="citation">Serván et al. (<a href="#ref-servan2018coexistence">2018</a>)</span>.</p>
<div id="a-trivial-case" class="section level3 hasAnchor" number="7.3.1">
<h3><span class="header-section-number">7.3.1</span> A trivial case<a href="ecological-assembly.html#a-trivial-case" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Suppose we take <span class="math inline">\(n\)</span> populations, each growing according to:</p>
<p><span class="math display">\[
\dot{x}_i = x_i (r_i - A_{ii} x_i)
\]</span></p>
<p>with <span class="math inline">\(A_{ii} &gt; 0\)</span>. We assume that the growth rates are sampled independently from a distribution: <span class="math inline">\(r_i \sim Q\)</span>, a distribution with mean <span class="math inline">\(0\)</span>, <span class="math inline">\(\mathbb E(r_i) = 0\)</span> and symmetric about its mean, such that <span class="math inline">\(\phi(r_i) = \phi(-r_i)\)</span>, where <span class="math inline">\(\phi(\cdot)\)</span> is the p.d.f. of distribution <span class="math inline">\(Q\)</span>. For example, we could choose <span class="math inline">\(r_i \sim \mathcal N(0, \sigma^2)\)</span>, or <span class="math inline">\(r_i \sim \mathcal U[-1, 1]\)</span>. Naturally, whenever <span class="math inline">\(r_i &gt; 0\)</span>, the population will grow to <span class="math inline">\(x_i^\star = r_i / A_{ii}\)</span>. If <span class="math inline">\(r_i &lt; 0\)</span>, on the other hand, the population will go extinct. Because of the symmetry around zero, each population has therefore probability <span class="math inline">\(1/2\)</span> of persisting. When we have <span class="math inline">\(n\)</span> populations, the number of coexisting populations <span class="math inline">\(\|k\|\)</span> is a random variable following the binomial distribution:</p>
<p><span class="math display">\[
p(\|k\| | n) = \binom{n}{\|k\|} \frac{1}{2^n}\quad \|k\| \sim \mathcal B\left(n, \frac{1}{2}\right)
\]</span></p>
</div>
<div id="the-random-zoo" class="section level3 hasAnchor" number="7.3.2">
<h3><span class="header-section-number">7.3.2</span> The random zoo<a href="ecological-assembly.html#the-random-zoo" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now consider a GLV model with random parameters. We choose <span class="math inline">\(r_i\)</span> from a distribution with mean zero and symmetric about the mean, and each interaction <span class="math inline">\(A_{ij}\)</span> from a distribution with mean zero and symmetric about the mean. Moreover, we add a constant to the diagonal of the matrix such that the symmetric part of <span class="math inline">\(A\)</span>, <span class="math inline">\(H(A) = \frac{1}{2} (A + A^T)\)</span> is positive definite. As we have seen before, this means that dynamics will always converge to a saturated equilibrium.</p>
<p>We now introduce a family of <span class="math inline">\(2^n\)</span> matrices <span class="math inline">\(S_i\)</span>, with:</p>
<p><span class="math display">\[
\begin{aligned}
S_1 &amp;= \begin{pmatrix}
1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 1 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
\vdots &amp; \vdots &amp; \vdots &amp;\ddots &amp;\vdots\\
0 &amp; 0 &amp; 0 &amp; 0 &amp;1
\end{pmatrix} = I_n\\
S_2 &amp;= \begin{pmatrix}
-1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 1 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 0 &amp; 1 &amp; \cdots &amp; 0\\
\vdots &amp; \vdots &amp; \vdots &amp;\ddots &amp;\vdots\\
0 &amp; 0 &amp; 0 &amp; 0 &amp;1
\end{pmatrix} \\
S_3 &amp;= \begin{pmatrix}
1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; -1 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 0 &amp; 1 &amp; \cdots &amp; 0\\
\vdots &amp; \vdots &amp; \vdots &amp;\ddots &amp;\vdots\\
0 &amp; 0 &amp; 0 &amp; 0 &amp;1
\end{pmatrix} \\
S_4 &amp;= \begin{pmatrix}
-1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; -1 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 0 &amp; 1 &amp; \cdots &amp; 0\\
\vdots &amp; \vdots &amp; \vdots &amp;\ddots &amp;\vdots\\
0 &amp; 0 &amp; 0 &amp; 0 &amp;1
\end{pmatrix}\\
\vdots\\
S_{2^n} &amp;= \begin{pmatrix}
-1 &amp; 0 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; -1 &amp; 0 &amp; \cdots &amp; 0\\
0 &amp; 0 &amp; -1 &amp; \cdots &amp; 0\\
\vdots &amp; \vdots &amp; \vdots &amp;\ddots &amp;\vdots\\
0 &amp; 0 &amp; 0 &amp; 0 &amp;-1
\end{pmatrix} = -I_n\\
\end{aligned}
\]</span></p>
<p>These matrices, called <em>signature matrices</em> are their own inverse (a matrix with this property is called <em>involutory</em>), and therefore, we have that <span class="math inline">\(S_i S_i = S_i^2 = I_n\)</span>. Geometrically, these matrices represent a reflection in each of the axes corresponding to the negated rows or columns.</p>
<p>We then consider what happens when we transform the parameters by multiplying them by <span class="math inline">\(S_i\)</span>. Take an arbitrary GLV system, and solve for the <span class="math inline">\(x\)</span> that makes <span class="math inline">\(r - Ax = 0_n\)</span>:</p>
<p><span class="math display">\[
x = A^{-1} r
\]</span></p>
<p>The vector is a feasible equilibrium for the system if <span class="math inline">\(x &gt; 0_n\)</span>. Now consider the transformed system:</p>
<p><span class="math display">\[
r&#39; = S_i r \quad A&#39; = S_iAS_i
\]</span></p>
<p>Note that <span class="math inline">\(A&#39;\)</span> is similar to <span class="math inline">\(A\)</span>, and therefore has the same eigenvalues; moreover, <span class="math inline">\(\phi(r&#39;) = \phi(r)\)</span> and <span class="math inline">\(\phi(A&#39;) = \phi(A)\)</span>, because of symmetry. Now compute the new solution</p>
<p><span class="math display">\[
\begin{aligned}
x&#39; &amp;= (A&#39;)^{-1} r&#39;\\
&amp;= (S_iAS_i)^{-1} S_i r\\
&amp;=S_iA^{-1}S_i S_i r\\
&amp;=S_iA^{-1}r\\
&amp;= S_i x
\end{aligned}
\]</span></p>
<p>That is, the transformation changes the sign of the components of <span class="math inline">\(x\)</span> corresponding to the negative coefficients in <span class="math inline">\(S_i\)</span>. Hence, for any possible choice of <span class="math inline">\(r\)</span> and <span class="math inline">\(A\)</span>, there is only one possible choice of <span class="math inline">\(S_i\)</span> that makes the equilibrium feasible. Because all matrices <span class="math inline">\(A&#39;\)</span> and all vectors <span class="math inline">\(r&#39;\)</span> have the same probability/density, then the probability of having a positive <span class="math inline">\(x\)</span> is exactly <span class="math inline">\(1/2^n\)</span>. Thus, a system with random parameters and a strongly stable matrix of interactions has the same probability of feasibility as that of a system in which populations are not interacting!</p>
<p>We have seen that when <span class="math inline">\(A\)</span> is such that <span class="math inline">\(H(A)\)</span> is negative definite, dynamics always converge to a saturated equilibrium. We want to compute the probability of finding a saturated equilibrium, in which populations belonging to set <span class="math inline">\(k\)</span> coexist, and the remaining populations cannot invade when rare.</p>
<p>We thus divide <span class="math inline">\(x\)</span>, <span class="math inline">\(r\)</span> and <span class="math inline">\(A\)</span> as:</p>
<p><span class="math display">\[
x = \begin{pmatrix}
y\\
z
\end{pmatrix} \quad \tilde{x} = \begin{pmatrix}
y^\star\\
0_{\|n-k\|}
\end{pmatrix} \quad A = \begin{pmatrix}
\mathfrak A &amp; \mathfrak B\\
\mathfrak C &amp; \mathfrak D
\end{pmatrix}\quad r = \begin{pmatrix}
\mathfrak r\\
\mathfrak s
\end{pmatrix}
\]</span></p>
<p>We also divide <span class="math inline">\(S_i\)</span> into:</p>
<p><span class="math display">\[
S_i = \begin{pmatrix}
S_{i}^{(k)} &amp; 0_{\|k\|, \|n-k\|}\\
0_{\|n-k\|, \|n\|} &amp; S_{i}^{(n-k)}
\end{pmatrix}
\]</span></p>
<p>When we use the signature matrix, the vector <span class="math inline">\(\mathfrak s\)</span> and the matrix <span class="math inline">\(\mathfrak C\)</span> become:</p>
<p><span class="math display">\[
\mathfrak s&#39; = S_i^{(n-k)} \mathfrak s\quad \mathfrak C&#39; = S_i^{(n-k)} \mathfrak C S_i^{(k)}
\]</span></p>
<p>The feasibility and nonivasibility condition, for a given choice of <span class="math inline">\(S_i\)</span> read:</p>
<p><span class="math display">\[
\begin{cases}
y&#39; = S_i^{(k)} (\mathfrak A^{-1}\mathfrak r) &gt; 0_{\|k\|}\\
D(z)^{-1} z =S_i^{(n-k)}(\mathfrak s + \mathfrak C y&#39;) &lt; 0_{\|n-k\|}\\
\end{cases}
\]</span></p>
<p>There is only one choice out of <span class="math inline">\(2^{\|k\|}\)</span> that makes <span class="math inline">\(y&#39; &gt; 0_{\|k\|}\)</span>; moreover, there is only one choice of <span class="math inline">\(S_i^{(n-k)}\)</span> that fulfills the noninvasibility condition, out of <span class="math inline">\(2^{\|n-k\|}\)</span> equiprobable choices. Then, we have that the probability of finding exactly <span class="math inline">\(\|k\|\)</span> populations coexisting at the saturated equilibrium, with the remaining <span class="math inline">\(\|n-k\|\)</span> that cannot invade is exactly <span class="math inline">\(1/(2^{\|k\|} \cdot 2^{\|n-k\|}) = 1/2^n\)</span>. The number of coexisting populations at the saturated equilibrium follows the binomial distribution:</p>
<p><span class="math display">\[
p(\|k\| | n) = \binom{n}{\|k\|} \frac{1}{2^n}\quad \|k\| \sim \mathcal B\left(n, \frac{1}{2}\right)
\]</span></p>
<p>exactly as for the non-interacting case.</p>
</div>
<div id="simulating-final-composition" class="section level3 hasAnchor" number="7.3.3">
<h3><span class="header-section-number">7.3.3</span> Simulating final composition<a href="ecological-assembly.html#simulating-final-composition" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This section details how the problem above can be simulated efficiently. First, we load some functions to integrate GLV dynamics and plot results:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="ecological-assembly.html#cb20-1" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/general_code_assembly.R&quot;</span>)</span></code></pre></div>
<p>And then write functions to give us a (barely) stable, symmetric matrix <span class="math inline">\(A\)</span>, and random growth rates:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="ecological-assembly.html#cb21-1" tabindex="-1"></a><span class="co"># function to build symmetric, Lyapunov Diagonally-stable matrix</span></span>
<span id="cb21-2"><a href="ecological-assembly.html#cb21-2" tabindex="-1"></a>build_LDstable <span class="ot">&lt;-</span> <span class="cf">function</span>(n){</span>
<span id="cb21-3"><a href="ecological-assembly.html#cb21-3" tabindex="-1"></a>  A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, n, n)</span>
<span id="cb21-4"><a href="ecological-assembly.html#cb21-4" tabindex="-1"></a>  A[<span class="fu">upper.tri</span>(A)] <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(n <span class="sc">*</span> (n <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">/</span> <span class="dv">2</span>)</span>
<span id="cb21-5"><a href="ecological-assembly.html#cb21-5" tabindex="-1"></a>  <span class="co"># make symmetric</span></span>
<span id="cb21-6"><a href="ecological-assembly.html#cb21-6" tabindex="-1"></a>  A <span class="ot">&lt;-</span> A <span class="sc">+</span> <span class="fu">t</span>(A)</span>
<span id="cb21-7"><a href="ecological-assembly.html#cb21-7" tabindex="-1"></a>  <span class="co"># now find the largest eigenvalue</span></span>
<span id="cb21-8"><a href="ecological-assembly.html#cb21-8" tabindex="-1"></a>  l1A <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">eigen</span>(A, <span class="at">only.values =</span> <span class="cn">TRUE</span>, <span class="at">symmetric =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values)</span>
<span id="cb21-9"><a href="ecological-assembly.html#cb21-9" tabindex="-1"></a>  <span class="cf">if</span> (l1A <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb21-10"><a href="ecological-assembly.html#cb21-10" tabindex="-1"></a>    <span class="co"># set the diagonal to make it stable</span></span>
<span id="cb21-11"><a href="ecological-assembly.html#cb21-11" tabindex="-1"></a>    <span class="fu">diag</span>(A) <span class="ot">&lt;-</span> <span class="fu">diag</span>(A) <span class="sc">-</span> l1A <span class="sc">-</span> <span class="fl">0.01</span></span>
<span id="cb21-12"><a href="ecological-assembly.html#cb21-12" tabindex="-1"></a>  }</span>
<span id="cb21-13"><a href="ecological-assembly.html#cb21-13" tabindex="-1"></a>  <span class="fu">return</span>(A)</span>
<span id="cb21-14"><a href="ecological-assembly.html#cb21-14" tabindex="-1"></a>}</span>
<span id="cb21-15"><a href="ecological-assembly.html#cb21-15" tabindex="-1"></a><span class="co"># function to get random growth rates</span></span>
<span id="cb21-16"><a href="ecological-assembly.html#cb21-16" tabindex="-1"></a>build_randomr <span class="ot">&lt;-</span> <span class="cf">function</span>(n){</span>
<span id="cb21-17"><a href="ecological-assembly.html#cb21-17" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb21-18"><a href="ecological-assembly.html#cb21-18" tabindex="-1"></a>}</span></code></pre></div>
<p>Now, we build a random system with seven species, and integrate the dynamics:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="ecological-assembly.html#cb22-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5</span>) <span class="co"># for reproducibility</span></span>
<span id="cb22-2"><a href="ecological-assembly.html#cb22-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">7</span></span>
<span id="cb22-3"><a href="ecological-assembly.html#cb22-3" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_LDstable</span>(n)</span>
<span id="cb22-4"><a href="ecological-assembly.html#cb22-4" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">build_randomr</span>(n)</span>
<span id="cb22-5"><a href="ecological-assembly.html#cb22-5" tabindex="-1"></a>x0 <span class="ot">&lt;-</span> <span class="fu">runif</span>(n)</span>
<span id="cb22-6"><a href="ecological-assembly.html#cb22-6" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">GLV_dynamics</span>(<span class="at">z0 =</span> x0, <span class="at">A =</span> A, <span class="at">r =</span> r, <span class="at">maxtime =</span> <span class="dv">25</span>, <span class="at">bytime =</span> <span class="fl">0.1</span>)</span>
<span id="cb22-7"><a href="ecological-assembly.html#cb22-7" tabindex="-1"></a><span class="fu">show</span>(<span class="fu">plot_dynamics</span>(out))</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/rndzooex1-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="ecological-assembly.html#cb23-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(</span>
<span id="cb23-2"><a href="ecological-assembly.html#cb23-2" tabindex="-1"></a>out<span class="sc">$</span>ts <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a href="ecological-assembly.html#cb23-3" tabindex="-1"></a>  <span class="fu">filter</span>(out<span class="sc">$</span>ts<span class="sc">$</span>time <span class="sc">==</span> <span class="fu">max</span>(out<span class="sc">$</span>ts<span class="sc">$</span>time))</span>
<span id="cb23-4"><a href="ecological-assembly.html#cb23-4" tabindex="-1"></a>)</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">time</th>
<th align="left">species</th>
<th align="right">density</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">25</td>
<td align="left">x1</td>
<td align="right">0.4290121</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x2</td>
<td align="right">0.2761340</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x3</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x4</td>
<td align="right">0.4021960</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x5</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x6</td>
<td align="right">0.6452999</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x7</td>
<td align="right">0.5776737</td>
</tr>
</tbody>
</table>
<p>As you can see, in this case, two species (3 and 5) go extinct, while the other ones reach a feasible equilibrium. Let’s try to start the system with species 3 and 5 at high abundance, to show that the equilibrium is indeed globally stable:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="ecological-assembly.html#cb24-1" tabindex="-1"></a>x0[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb24-2"><a href="ecological-assembly.html#cb24-2" tabindex="-1"></a>x0[<span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb24-3"><a href="ecological-assembly.html#cb24-3" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">GLV_dynamics</span>(<span class="at">z0 =</span> x0, <span class="at">A =</span> A, <span class="at">r =</span> r, <span class="at">maxtime =</span> <span class="dv">25</span>, <span class="at">bytime =</span> <span class="fl">0.1</span>)</span>
<span id="cb24-4"><a href="ecological-assembly.html#cb24-4" tabindex="-1"></a><span class="fu">show</span>(<span class="fu">plot_dynamics</span>(out))</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/rndzooex2-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="ecological-assembly.html#cb25-1" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(</span>
<span id="cb25-2"><a href="ecological-assembly.html#cb25-2" tabindex="-1"></a>out<span class="sc">$</span>ts <span class="sc">%&gt;%</span> </span>
<span id="cb25-3"><a href="ecological-assembly.html#cb25-3" tabindex="-1"></a>  <span class="fu">filter</span>(out<span class="sc">$</span>ts<span class="sc">$</span>time <span class="sc">==</span> <span class="fu">max</span>(out<span class="sc">$</span>ts<span class="sc">$</span>time))</span>
<span id="cb25-4"><a href="ecological-assembly.html#cb25-4" tabindex="-1"></a>)</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">time</th>
<th align="left">species</th>
<th align="right">density</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">25</td>
<td align="left">x1</td>
<td align="right">0.4291312</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x2</td>
<td align="right">0.2759277</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x3</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x4</td>
<td align="right">0.4020638</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x5</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="right">25</td>
<td align="left">x6</td>
<td align="right">0.6454349</td>
</tr>
<tr class="odd">
<td align="right">25</td>
<td align="left">x7</td>
<td align="right">0.5777422</td>
</tr>
</tbody>
</table>
<p>As you can see, the system still goes to the same equilibrium, with 3 and 5 extinct. Turns out, for this type of system, one does not even need to integrate dynamics: the Lemke–Howson algorithm can be adapted to solve the problem efficiently <span class="citation">(<a href="#ref-servan2018coexistence">Serván et al. 2018</a>)</span> (this is an efficient algorithm to find Nash equilibria in a game). I have coded up the algorithm already, and you can load the function <code>get_final_composition(A, r)</code> by typing:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="ecological-assembly.html#cb26-1" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/L-H.R&quot;</span>)</span>
<span id="cb26-2"><a href="ecological-assembly.html#cb26-2" tabindex="-1"></a><span class="fu">get_final_composition</span>(A, r) <span class="co"># use LH instead of integrating dynamics</span></span></code></pre></div>
<p>[1] 0.4290300 0.2761030 0.0000000 0.4021761 0.0000000 0.6453202 0.5776840</p>
<p>To simulate the random zoo, we take random matrices and random growth rates with <span class="math inline">\(n\)</span> species, and tally the number of coexisting species over several simulations:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="ecological-assembly.html#cb27-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb27-2"><a href="ecological-assembly.html#cb27-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb27-3"><a href="ecological-assembly.html#cb27-3" tabindex="-1"></a>nsim <span class="ot">&lt;-</span> <span class="dv">2000</span></span>
<span id="cb27-4"><a href="ecological-assembly.html#cb27-4" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">simulation =</span> <span class="dv">1</span><span class="sc">:</span>nsim, </span>
<span id="cb27-5"><a href="ecological-assembly.html#cb27-5" tabindex="-1"></a>                      <span class="at">ncoexisting =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim))</span>
<span id="cb27-6"><a href="ecological-assembly.html#cb27-6" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nsim){</span>
<span id="cb27-7"><a href="ecological-assembly.html#cb27-7" tabindex="-1"></a>  <span class="co"># build the matrix and vector</span></span>
<span id="cb27-8"><a href="ecological-assembly.html#cb27-8" tabindex="-1"></a>  A <span class="ot">&lt;-</span> <span class="fu">build_LDstable</span>(n)</span>
<span id="cb27-9"><a href="ecological-assembly.html#cb27-9" tabindex="-1"></a>  r <span class="ot">&lt;-</span> <span class="fu">build_randomr</span>(n)</span>
<span id="cb27-10"><a href="ecological-assembly.html#cb27-10" tabindex="-1"></a>  xstar <span class="ot">&lt;-</span> <span class="fu">get_final_composition</span>(A, r)</span>
<span id="cb27-11"><a href="ecological-assembly.html#cb27-11" tabindex="-1"></a>  results<span class="sc">$</span>ncoexisting[i] <span class="ot">&lt;-</span> <span class="fu">sum</span>(xstar <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb27-12"><a href="ecological-assembly.html#cb27-12" tabindex="-1"></a>}</span>
<span id="cb27-13"><a href="ecological-assembly.html#cb27-13" tabindex="-1"></a>pl <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> results) <span class="sc">+</span> </span>
<span id="cb27-14"><a href="ecological-assembly.html#cb27-14" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> ncoexisting) <span class="sc">+</span> </span>
<span id="cb27-15"><a href="ecological-assembly.html#cb27-15" tabindex="-1"></a>  <span class="fu">geom_bar</span>() <span class="sc">+</span> </span>
<span id="cb27-16"><a href="ecological-assembly.html#cb27-16" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="st">&quot;number of species coexisting&quot;</span>, <span class="at">breaks =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb27-17"><a href="ecological-assembly.html#cb27-17" tabindex="-1"></a><span class="fu">show</span>(pl)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/randomzoo-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="ecological-assembly.html#cb28-1" tabindex="-1"></a><span class="co"># add binomial distribution</span></span>
<span id="cb28-2"><a href="ecological-assembly.html#cb28-2" tabindex="-1"></a>tbinom <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ncoexisting =</span> <span class="dv">0</span><span class="sc">:</span>n, </span>
<span id="cb28-3"><a href="ecological-assembly.html#cb28-3" tabindex="-1"></a>                 <span class="at">expectation =</span> nsim <span class="sc">*</span> <span class="fu">dbinom</span>(<span class="dv">0</span><span class="sc">:</span>n, n, <span class="fl">0.5</span>))</span>
<span id="cb28-4"><a href="ecological-assembly.html#cb28-4" tabindex="-1"></a>pl <span class="ot">&lt;-</span> pl <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">data =</span> tbinom, <span class="fu">aes</span>(<span class="at">x =</span> ncoexisting, <span class="at">y =</span> expectation))</span>
<span id="cb28-5"><a href="ecological-assembly.html#cb28-5" tabindex="-1"></a><span class="fu">show</span>(pl)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/randomzoo-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>We find again that the number of coexisting species follows the binomial distribution with parameters <span class="math inline">\(n\)</span> and <span class="math inline">\(1/2\)</span> (i.e., exactly the same result found for non interacting species).</p>
</div>
<div id="a-random-zoo-in-the-wild" class="section level3 hasAnchor" number="7.3.4">
<h3><span class="header-section-number">7.3.4</span> A random zoo in the wild<a href="ecological-assembly.html#a-random-zoo-in-the-wild" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One might object that the derivations above have no parallel in reality. However, they can be seen as a simplified version of an experiment done routinely by ecologists. Take for example the work of <span class="citation">Bittleston et al. (<a href="#ref-bittleston2020context">2020</a>)</span>. They sampled bacterial communities from ten different pitcher plants, and cultured them in the laboratory on a synthetic medium. They tracked the change of richness in time. This is their Figure 2 (look at panel b):</p>
<center>
<img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41467-020-15169-0/MediaObjects/41467_2020_15169_Fig2_HTML.png">
</center>
<p>The more the species present at day 3 (i.e., once those with a negative growth rate had disappeared), the more one would find after two months—and the relationship is linear!</p>
</div>
</div>
<div id="bottom-up-assembly" class="section level2 hasAnchor" number="7.4">
<h2><span class="header-section-number">7.4</span> Bottom-up assembly<a href="ecological-assembly.html#bottom-up-assembly" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote>
<p>Mother Nature, of course, does not assemble her networks by throwing <em>n</em> species together in one go. It makes more sense to assume that she adds one species after another through successive invasions.</p>
<p><span class="citation">Sigmund (<a href="#ref-sigmuiud1995darwin">1995</a>)</span></p>
</blockquote>
<p>Having considered the case in which all species are thrown into the habitat at the same time (<strong>top-down</strong> assembly), we consider a process in which we start from the “bare ground” and build our community from the <strong>bottom-up</strong>.</p>
<p>Note that in top-down assembly, any feasible equilibrium can be achieved by starting with the appropriate initial conditions; being slightly less generous, we can think of being able to assemble from the top-down any “persistent” (e.g., stable), feasible community we can form from the pool. It makes therefore sense to ask whether these same states can or cannot be accessed when assembling the community from the ground up.</p>
<div id="simulating-bottom-up-assembly" class="section level3 hasAnchor" number="7.4.1">
<h3><span class="header-section-number">7.4.1</span> Simulating bottom-up assembly<a href="ecological-assembly.html#simulating-bottom-up-assembly" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now let’s make it more complicated: we assemble an ecological community from the ground up. At each step, we introduce a species at low abundance, starting from an empty community. We then compute the new equilibrium, completing a step of the assembly:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="ecological-assembly.html#cb29-1" tabindex="-1"></a>assembly_one_step <span class="ot">&lt;-</span> <span class="cf">function</span>(x, r, A){</span>
<span id="cb29-2"><a href="ecological-assembly.html#cb29-2" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(A)</span>
<span id="cb29-3"><a href="ecological-assembly.html#cb29-3" tabindex="-1"></a>  invader <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n, <span class="dv">1</span>)</span>
<span id="cb29-4"><a href="ecological-assembly.html#cb29-4" tabindex="-1"></a>  x[invader] <span class="ot">&lt;-</span> <span class="fl">0.001</span> <span class="co"># introduce the invader at low abundance</span></span>
<span id="cb29-5"><a href="ecological-assembly.html#cb29-5" tabindex="-1"></a>  present <span class="ot">&lt;-</span> x <span class="sc">&gt;</span> <span class="dv">0</span> <span class="co"># these are the species present now</span></span>
<span id="cb29-6"><a href="ecological-assembly.html#cb29-6" tabindex="-1"></a>  <span class="co"># compute new equilibrium</span></span>
<span id="cb29-7"><a href="ecological-assembly.html#cb29-7" tabindex="-1"></a>  y <span class="ot">&lt;-</span> <span class="fu">get_final_composition</span>(A[present, present, <span class="at">drop =</span> <span class="cn">FALSE</span>], r[present])</span>
<span id="cb29-8"><a href="ecological-assembly.html#cb29-8" tabindex="-1"></a>  x[present] <span class="ot">&lt;-</span> y</span>
<span id="cb29-9"><a href="ecological-assembly.html#cb29-9" tabindex="-1"></a>  <span class="fu">return</span>(x)</span>
<span id="cb29-10"><a href="ecological-assembly.html#cb29-10" tabindex="-1"></a>}</span></code></pre></div>
<p>Now we can take a species pool along with their parameters, and try to assemble the system until we can no longer add any more species (i.e., until we reach a saturated equilibrium):</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="ecological-assembly.html#cb30-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">7</span>)</span>
<span id="cb30-2"><a href="ecological-assembly.html#cb30-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb30-3"><a href="ecological-assembly.html#cb30-3" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_LDstable</span>(n)</span>
<span id="cb30-4"><a href="ecological-assembly.html#cb30-4" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">build_randomr</span>(n)</span>
<span id="cb30-5"><a href="ecological-assembly.html#cb30-5" tabindex="-1"></a><span class="co"># start with no species</span></span>
<span id="cb30-6"><a href="ecological-assembly.html#cb30-6" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, n)</span>
<span id="cb30-7"><a href="ecological-assembly.html#cb30-7" tabindex="-1"></a><span class="co"># assemble for 40 steps and keep track of richness and composition</span></span>
<span id="cb30-8"><a href="ecological-assembly.html#cb30-8" tabindex="-1"></a>ninvasions <span class="ot">&lt;-</span> <span class="dv">40</span></span>
<span id="cb30-9"><a href="ecological-assembly.html#cb30-9" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">invasion =</span> <span class="dv">1</span><span class="sc">:</span>ninvasions, </span>
<span id="cb30-10"><a href="ecological-assembly.html#cb30-10" tabindex="-1"></a>                  <span class="at">richness =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ninvasions),</span>
<span id="cb30-11"><a href="ecological-assembly.html#cb30-11" tabindex="-1"></a>                  <span class="at">composition =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ninvasions))</span>
<span id="cb30-12"><a href="ecological-assembly.html#cb30-12" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>ninvasions){</span>
<span id="cb30-13"><a href="ecological-assembly.html#cb30-13" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">assembly_one_step</span>(x, r, A) </span>
<span id="cb30-14"><a href="ecological-assembly.html#cb30-14" tabindex="-1"></a>  results<span class="sc">$</span>richness[i] <span class="ot">&lt;-</span> <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb30-15"><a href="ecological-assembly.html#cb30-15" tabindex="-1"></a>  results<span class="sc">$</span>composition[i] <span class="ot">&lt;-</span> <span class="fu">paste</span>((<span class="dv">1</span><span class="sc">:</span>n)[x<span class="sc">&gt;</span><span class="dv">0</span>], <span class="at">collapse =</span> <span class="st">&quot;-&quot;</span>)</span>
<span id="cb30-16"><a href="ecological-assembly.html#cb30-16" tabindex="-1"></a>}</span>
<span id="cb30-17"><a href="ecological-assembly.html#cb30-17" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(results))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">invasion</th>
<th align="right">richness</th>
<th align="left">composition</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">0</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="right">0</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="right">1</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">2</td>
<td align="left">1-2</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">2</td>
<td align="left">1-2</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="right">2</td>
<td align="left">1-2</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="ecological-assembly.html#cb31-1" tabindex="-1"></a>pl <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(results) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> invasion, <span class="at">y =</span> richness, <span class="at">label =</span> composition) <span class="sc">+</span> </span>
<span id="cb31-2"><a href="ecological-assembly.html#cb31-2" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">nudge_y =</span> <span class="fl">0.25</span>, <span class="at">angle =</span> <span class="dv">90</span>)<span class="sc">+</span> </span>
<span id="cb31-3"><a href="ecological-assembly.html#cb31-3" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, n <span class="sc">*</span> <span class="fl">1.25</span>))</span>
<span id="cb31-4"><a href="ecological-assembly.html#cb31-4" tabindex="-1"></a><span class="fu">show</span>(pl)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/rndzooassembly-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Now let’s roll back history and assemble again:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="ecological-assembly.html#cb32-1" tabindex="-1"></a><span class="co"># start with no species</span></span>
<span id="cb32-2"><a href="ecological-assembly.html#cb32-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, n)</span>
<span id="cb32-3"><a href="ecological-assembly.html#cb32-3" tabindex="-1"></a><span class="co"># assemble for 40 steps and keep track of richness</span></span>
<span id="cb32-4"><a href="ecological-assembly.html#cb32-4" tabindex="-1"></a>ninvasions <span class="ot">&lt;-</span> <span class="dv">40</span></span>
<span id="cb32-5"><a href="ecological-assembly.html#cb32-5" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">invasion =</span> <span class="dv">1</span><span class="sc">:</span>ninvasions, </span>
<span id="cb32-6"><a href="ecological-assembly.html#cb32-6" tabindex="-1"></a>                  <span class="at">richness =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ninvasions),</span>
<span id="cb32-7"><a href="ecological-assembly.html#cb32-7" tabindex="-1"></a>                  <span class="at">composition =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ninvasions))</span>
<span id="cb32-8"><a href="ecological-assembly.html#cb32-8" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>ninvasions){</span>
<span id="cb32-9"><a href="ecological-assembly.html#cb32-9" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">assembly_one_step</span>(x, r, A) </span>
<span id="cb32-10"><a href="ecological-assembly.html#cb32-10" tabindex="-1"></a>  results<span class="sc">$</span>richness[i] <span class="ot">&lt;-</span> <span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb32-11"><a href="ecological-assembly.html#cb32-11" tabindex="-1"></a>  results<span class="sc">$</span>composition[i] <span class="ot">&lt;-</span> <span class="fu">paste</span>((<span class="dv">1</span><span class="sc">:</span>n)[x<span class="sc">&gt;</span><span class="dv">0</span>], <span class="at">collapse =</span> <span class="st">&quot;-&quot;</span>)</span>
<span id="cb32-12"><a href="ecological-assembly.html#cb32-12" tabindex="-1"></a>}</span>
<span id="cb32-13"><a href="ecological-assembly.html#cb32-13" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(results))</span></code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">invasion</th>
<th align="right">richness</th>
<th align="left">composition</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">1</td>
<td align="left">4</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="right">1</td>
<td align="left">4</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="right">1</td>
<td align="left">4</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">1</td>
<td align="left">4</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">2</td>
<td align="left">4-10</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="right">2</td>
<td align="left">4-10</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="ecological-assembly.html#cb33-1" tabindex="-1"></a>pl <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(results) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> invasion, <span class="at">y =</span> richness, <span class="at">label =</span> composition) <span class="sc">+</span> </span>
<span id="cb33-2"><a href="ecological-assembly.html#cb33-2" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">nudge_y =</span> <span class="fl">0.25</span>, <span class="at">angle =</span> <span class="dv">90</span>)<span class="sc">+</span> </span>
<span id="cb33-3"><a href="ecological-assembly.html#cb33-3" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, n <span class="sc">*</span> <span class="fl">1.25</span>))</span>
<span id="cb33-4"><a href="ecological-assembly.html#cb33-4" tabindex="-1"></a><span class="fu">show</span>(pl)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/alternativehistory-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>As you can see, despite taking a different assembly history, we reach the same final composition. In fact, this is exactly what we would expect if we were to throw all species in the environment at the same time:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="ecological-assembly.html#cb34-1" tabindex="-1"></a><span class="fu">get_final_composition</span>(A, r)</span></code></pre></div>
<p>[1] 1.7588344 2.5440743 0.0000000 0.7860178 1.8730535 2.5502055 0.6961259
[8] 1.6956878 0.0000000 0.0000000</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="ecological-assembly.html#cb35-1" tabindex="-1"></a>x</span></code></pre></div>
<p>[1] 1.7588344 2.5440743 0.0000000 0.7860178 1.8730535 2.5502055 0.6961259
[8] 1.6956878 0.0000000 0.0000000</p>
<p>In fact, given enough time, any assembly history for a symmetric, stable matrix <span class="math inline">\(A\)</span> will eventually reach the final composition represented by the saturated equilibrium (<span class="citation">Serván and Allesina (<a href="#ref-servan2020tractable">2020</a>)</span>).</p>
<p>Interestingly, this is not the case when the matrix is not symmetric. <span class="citation">Serván et al. (<a href="#ref-servan2018coexistence">2018</a>)</span> conjectured however that the probability of finding a system whose final composition cannot be assembled one species at a time decreases rapidly with the size of the pool, as long as <span class="math inline">\(A\)</span> is Lyapunov Diagonally stable.</p>
</div>
<div id="network-spandrels" class="section level3 hasAnchor" number="7.4.2">
<h3><span class="header-section-number">7.4.2</span> Network spandrels<a href="ecological-assembly.html#network-spandrels" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>What is the network structure of the assembled community vs. that of the initial pool? Can we detect a signature of the forces acting on the community such that some species can persist, while other go extinct? To answer these questions, we start by considering a larger pool of species:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="ecological-assembly.html#cb36-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb36-2"><a href="ecological-assembly.html#cb36-2" tabindex="-1"></a><span class="co"># initial pool</span></span>
<span id="cb36-3"><a href="ecological-assembly.html#cb36-3" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_LDstable</span>(<span class="dv">200</span>)</span>
<span id="cb36-4"><a href="ecological-assembly.html#cb36-4" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">build_randomr</span>(<span class="dv">200</span>)</span>
<span id="cb36-5"><a href="ecological-assembly.html#cb36-5" tabindex="-1"></a><span class="co"># final composition</span></span>
<span id="cb36-6"><a href="ecological-assembly.html#cb36-6" tabindex="-1"></a>xstar <span class="ot">&lt;-</span> <span class="fu">get_final_composition</span>(A, r)</span>
<span id="cb36-7"><a href="ecological-assembly.html#cb36-7" tabindex="-1"></a>A_pruned <span class="ot">&lt;-</span> A[xstar <span class="sc">&gt;</span> <span class="dv">0</span>, xstar <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb36-8"><a href="ecological-assembly.html#cb36-8" tabindex="-1"></a>r_pruned <span class="ot">&lt;-</span> r[xstar <span class="sc">&gt;</span> <span class="dv">0</span>]</span></code></pre></div>
<p>Are the properties of matrix <span class="math inline">\(A\)</span> different from those of the pruned version <span class="math inline">\(\tilde{A}\)</span>? We probe this in two ways: first, we build a graph with the strongest interactions, and plot it.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="ecological-assembly.html#cb37-1" tabindex="-1"></a><span class="co"># draw an edge i-&gt;j if Bij is above the 75% of strengths (in abs)</span></span>
<span id="cb37-2"><a href="ecological-assembly.html#cb37-2" tabindex="-1"></a>plot_graph_strong <span class="ot">&lt;-</span> <span class="cf">function</span>(B, <span class="at">quantile =</span> <span class="fl">0.75</span>){</span>
<span id="cb37-3"><a href="ecological-assembly.html#cb37-3" tabindex="-1"></a>  Bstrong <span class="ot">&lt;-</span> <span class="fu">abs</span>(B)</span>
<span id="cb37-4"><a href="ecological-assembly.html#cb37-4" tabindex="-1"></a>  <span class="fu">diag</span>(Bstrong) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb37-5"><a href="ecological-assembly.html#cb37-5" tabindex="-1"></a>  Bstrong[Bstrong <span class="sc">&lt;</span> <span class="fu">quantile</span>(Bstrong, quantile)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb37-6"><a href="ecological-assembly.html#cb37-6" tabindex="-1"></a>  gr <span class="ot">&lt;-</span> <span class="fu">graph_from_adjacency_matrix</span>((Bstrong <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">*</span> <span class="dv">1</span>, <span class="at">mode =</span> <span class="st">&quot;undirected&quot;</span>)</span>
<span id="cb37-7"><a href="ecological-assembly.html#cb37-7" tabindex="-1"></a>  <span class="fu">plot</span>(gr, <span class="at">vertex.size=</span><span class="dv">10</span>, <span class="at">vertex.label=</span><span class="cn">NA</span>, <span class="at">layout=</span>layout_with_fr)</span>
<span id="cb37-8"><a href="ecological-assembly.html#cb37-8" tabindex="-1"></a>}</span>
<span id="cb37-9"><a href="ecological-assembly.html#cb37-9" tabindex="-1"></a><span class="fu">plot_graph_strong</span>(A)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/plotnetworksrz-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="ecological-assembly.html#cb38-1" tabindex="-1"></a><span class="fu">plot_graph_strong</span>(A_pruned)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/plotnetworksrz-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>There seems to be no special structure. A more powerful way to show the same is to plot the eigenvalues of <span class="math inline">\(A\)</span> and <span class="math inline">\(\tilde{A}\)</span>. For a symmetric matrix with off-diagonal elements centered at zero, the eigenvalues should follow Wigner’s semicircle law:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="ecological-assembly.html#cb39-1" tabindex="-1"></a>plot_eigen <span class="ot">&lt;-</span> <span class="cf">function</span>(B){</span>
<span id="cb39-2"><a href="ecological-assembly.html#cb39-2" tabindex="-1"></a>  evals <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">lambda =</span> <span class="fu">eigen</span>(B, <span class="at">only.values =</span> <span class="cn">TRUE</span>, <span class="at">symmetric =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values)</span>
<span id="cb39-3"><a href="ecological-assembly.html#cb39-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> evals) <span class="sc">+</span> </span>
<span id="cb39-4"><a href="ecological-assembly.html#cb39-4" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> lambda) <span class="sc">+</span> </span>
<span id="cb39-5"><a href="ecological-assembly.html#cb39-5" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="fu">as.integer</span>(<span class="fu">nrow</span>(evals) <span class="sc">/</span> <span class="dv">10</span>), <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb39-6"><a href="ecological-assembly.html#cb39-6" tabindex="-1"></a>}</span>
<span id="cb39-7"><a href="ecological-assembly.html#cb39-7" tabindex="-1"></a><span class="fu">plot_eigen</span>(A)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/ploteigenrz-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="ecological-assembly.html#cb40-1" tabindex="-1"></a><span class="fu">plot_eigen</span>(A_pruned)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/ploteigenrz-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>As such, the matrix of interactions before/after dynamics seem to have the same properties. However, as shown in <span class="citation">Serván et al. (<a href="#ref-servan2018coexistence">2018</a>)</span>, the distribution of the growth rates changes in a non-trivial way:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="ecological-assembly.html#cb41-1" tabindex="-1"></a>toplot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">r =</span> r, <span class="at">type =</span> <span class="st">&quot;before&quot;</span>)</span>
<span id="cb41-2"><a href="ecological-assembly.html#cb41-2" tabindex="-1"></a>toplot <span class="ot">&lt;-</span> <span class="fu">rbind</span>(toplot, <span class="fu">data.frame</span>(<span class="at">r =</span> r_pruned, <span class="at">type =</span> <span class="st">&quot;after&quot;</span>))</span>
<span id="cb41-3"><a href="ecological-assembly.html#cb41-3" tabindex="-1"></a><span class="fu">ggplot</span>(toplot) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x =</span> r, <span class="at">fill =</span> type) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">position =</span> <span class="st">&quot;dodge&quot;</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/plotrbeforeafter-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>But what if species were to be related to each other? For example, suppose <span class="math inline">\(r_i = 1\)</span> for all species, and build a matrix in which the interactions of species <span class="math inline">\(i + 1\)</span> are obtained by mutating slightly those of species <span class="math inline">\(i\)</span>:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="ecological-assembly.html#cb42-1" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb42-2"><a href="ecological-assembly.html#cb42-2" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">200</span>)</span>
<span id="cb42-3"><a href="ecological-assembly.html#cb42-3" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, n, n)</span>
<span id="cb42-4"><a href="ecological-assembly.html#cb42-4" tabindex="-1"></a><span class="co"># set first species</span></span>
<span id="cb42-5"><a href="ecological-assembly.html#cb42-5" tabindex="-1"></a>A[<span class="dv">1</span>, <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb42-6"><a href="ecological-assembly.html#cb42-6" tabindex="-1"></a><span class="co"># now each species is obtained by mutating the previous one</span></span>
<span id="cb42-7"><a href="ecological-assembly.html#cb42-7" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>n){</span>
<span id="cb42-8"><a href="ecological-assembly.html#cb42-8" tabindex="-1"></a>  ai <span class="ot">&lt;-</span> A[i <span class="sc">-</span><span class="dv">1</span>,] <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.05</span> <span class="sc">*</span> <span class="fu">runif</span>(n))</span>
<span id="cb42-9"><a href="ecological-assembly.html#cb42-9" tabindex="-1"></a>  A[i, ] <span class="ot">&lt;-</span> A[i, ] <span class="sc">+</span> ai</span>
<span id="cb42-10"><a href="ecological-assembly.html#cb42-10" tabindex="-1"></a>  A[, i] <span class="ot">&lt;-</span> A[, i] <span class="sc">+</span> ai</span>
<span id="cb42-11"><a href="ecological-assembly.html#cb42-11" tabindex="-1"></a>  A[i, i] <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb42-12"><a href="ecological-assembly.html#cb42-12" tabindex="-1"></a>}</span>
<span id="cb42-13"><a href="ecological-assembly.html#cb42-13" tabindex="-1"></a><span class="co"># make LD-stable</span></span>
<span id="cb42-14"><a href="ecological-assembly.html#cb42-14" tabindex="-1"></a>l1 <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">eigen</span>(A, <span class="at">only.values =</span> <span class="cn">TRUE</span>, <span class="at">symmetric =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values)</span>
<span id="cb42-15"><a href="ecological-assembly.html#cb42-15" tabindex="-1"></a><span class="cf">if</span> (l1 <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="fu">diag</span>(A) <span class="ot">&lt;-</span> <span class="fu">diag</span>(A) <span class="sc">-</span> l1 <span class="sc">*</span> <span class="fl">1.01</span></span></code></pre></div>
<p>Now each species is similar to the previous one:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="ecological-assembly.html#cb43-1" tabindex="-1"></a><span class="fu">image</span>(A[<span class="dv">1</span><span class="sc">:</span>n, n<span class="sc">:</span><span class="dv">1</span>])</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/similarimage-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Perform the pruning, and plot networks and eigenvalues:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="ecological-assembly.html#cb44-1" tabindex="-1"></a>xstar <span class="ot">&lt;-</span> <span class="fu">get_final_composition</span>(A, r)</span>
<span id="cb44-2"><a href="ecological-assembly.html#cb44-2" tabindex="-1"></a>A_pruned <span class="ot">&lt;-</span> A[xstar <span class="sc">&gt;</span> <span class="dv">0</span>, xstar <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb44-3"><a href="ecological-assembly.html#cb44-3" tabindex="-1"></a>r_pruned <span class="ot">&lt;-</span> r[xstar <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb44-4"><a href="ecological-assembly.html#cb44-4" tabindex="-1"></a><span class="fu">plot_graph_strong</span>(A)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/structruredspandrels-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="ecological-assembly.html#cb45-1" tabindex="-1"></a><span class="fu">plot_graph_strong</span>(A_pruned)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/structruredspandrels-2.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="ecological-assembly.html#cb46-1" tabindex="-1"></a><span class="fu">plot_eigen</span>(A)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/structruredspandrels-3.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="ecological-assembly.html#cb47-1" tabindex="-1"></a><span class="fu">plot_eigen</span>(A_pruned)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/structruredspandrels-4.png" width="480" style="display: block; margin: auto;" /></p>
<p>Meaning that if our matrix <span class="math inline">\(A\)</span> is structured, we will recover a structured matrix after pruning, while if <span class="math inline">\(A\)</span> is unstructured, we will recover an unstructured matrix. <span class="citation">Maynard et al. (<a href="#ref-maynard2018network">2018</a>)</span> showed that a well-defined network structure could be a “network spandrel” (cfr. <span class="citation">Gould and Lewontin (<a href="#ref-gould1979spandrels">1979</a>)</span> and <span class="citation">Solé and Valverde (<a href="#ref-sole2006network">2006</a>)</span>) arising from the way new species are introduced, rather than a “signature of stability”.</p>
</div>
<div id="an-assembly-graph" class="section level3 hasAnchor" number="7.4.3">
<h3><span class="header-section-number">7.4.3</span> An assembly graph<a href="ecological-assembly.html#an-assembly-graph" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In GLV, a given (sub-)community has at most one feasible equilibrium; that is, there is <strong>no true multi-stability in GLV</strong>: we can find the system at different stable states, but <strong>they have to differ in composition</strong>. Because of this fact, we can devise a scheme to label the possible states our community can be in.</p>
<p>We call <span class="math inline">\(0\)</span> the state in which no species are present, <span class="math inline">\(1\)</span> the state in which only species 1 is present, <span class="math inline">\(2\)</span> the state in which only species 2 is present, <span class="math inline">\(3\)</span> the state in which species 1 and 2 are both present, and so on. Practically, we take the community composition to be the base-2 representation of the label. For example, label <span class="math inline">\(11\)</span> in a community of 6 species corresponds to <span class="math inline">\(001011\)</span> (i.e., a state in which species 1, 2, and 4 are present). As this notation makes obvious, for a given pool of <span class="math inline">\(n\)</span> species, we can have up to <span class="math inline">\(2^n - 1\)</span> feasible equilibria. As we saw in Lecture 1, the existence of a feasible equilibrium is a necessary (but not sufficient—we should require also some form of stability/permanence) condition for coexistence.</p>
<p>Clearly, any feasible (and persistent/stable) sub-community can be observed by initializing the system at (or, in case of locally/globally stable configurations, close to) the desired densities. On the other hand, unstable configurations will eventually collapse to some other sub-community. As such, we take the labels/states representing stable/persistent communities to be the <strong>nodes</strong> in a directed graph. Then, we take the <strong>edges</strong> of this graph to represent invasions, moving the local community from one state to another. To keep the graph simple, we only consider “successful” invasions (i.e., those for which the initial and final state differ), thereby removing the need for “self-loops”.</p>
<p>This <strong>assembly graph</strong> was considered several times in the literature (see for example <span class="citation">Law and Morton (<a href="#ref-law1993alternative">1993</a>)</span>, <span class="citation">Hang-Kwang and Pimm (<a href="#ref-hang1993assembly">1993</a>)</span>, <span class="citation">Schreiber and Rittenhouse (<a href="#ref-schreiber2004simple">2004</a>)</span>, <span class="citation">Capitán et al. (<a href="#ref-capitan2009statistical">2009</a>)</span>). Here, we follow the approach <span class="citation">Serván and Allesina (<a href="#ref-servan2020tractable">2020</a>)</span>, and note that the assembly graph fully describes the assembly process whenever the assumptions that we’ve made at the onset of our exploration (invasions are rare, invasions are small, dynamics converge to equilibria) are satisfied. When this is the case, we can study the assembly process in its full glory by studying a graph (which definitely sounds more fun!).</p>
</div>
<div id="how-many-invasions" class="section level3 hasAnchor" number="7.4.4">
<h3><span class="header-section-number">7.4.4</span> How many invasions?<a href="ecological-assembly.html#how-many-invasions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, we might want to think of the problem of invasion. The bottom-up assembly can be seen as a single, massive invasion. At the other extreme, we have assembly proceeding with invasions of a single species at a time. Of course, we can imagine anything in between: species invade in small groups, there is a distribution describing the number of species invading at each step, etc.</p>
<p>For example, let’s build the assembly graph for a given set of (random) parameters: we take <span class="math inline">\(A\)</span> to be a symmetric, <strong>stable</strong> nonpositive matrix (e.g., representing competition between species), and <span class="math inline">\(r\)</span> to be a vector of positive, random growth rates. Let’s build the assembly graph when we consider that species can enter the system only one at a time:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="ecological-assembly.html#cb48-1" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/general_code_assembly.R&quot;</span>)</span>
<span id="cb48-2"><a href="ecological-assembly.html#cb48-2" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/L-H.R&quot;</span>) <span class="co"># Lemke-Howson algorithm to get saturated equilibrium</span></span>
<span id="cb48-3"><a href="ecological-assembly.html#cb48-3" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/build_assembly_graph.R&quot;</span>)</span>
<span id="cb48-4"><a href="ecological-assembly.html#cb48-4" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/build_assembly_graph_unstable.R&quot;</span>) <span class="co"># code to build and draw assembly graphs for symmetric matrices</span></span>
<span id="cb48-5"><a href="ecological-assembly.html#cb48-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4</span>) <span class="co"># for reproducibility</span></span>
<span id="cb48-6"><a href="ecological-assembly.html#cb48-6" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_competitive_stable</span>(<span class="dv">4</span>)</span>
<span id="cb48-7"><a href="ecological-assembly.html#cb48-7" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">4</span>)</span>
<span id="cb48-8"><a href="ecological-assembly.html#cb48-8" tabindex="-1"></a>assembly_invasion_1 <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A)</span>
<span id="cb48-9"><a href="ecological-assembly.html#cb48-9" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(assembly_invasion_1<span class="sc">$</span>graph, assembly_invasion_1<span class="sc">$</span>info)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-55-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>What if we allow for two invasions at a time?</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="ecological-assembly.html#cb49-1" tabindex="-1"></a>assembly_invasion_2 <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A, <span class="dv">2</span>)</span>
<span id="cb49-2"><a href="ecological-assembly.html#cb49-2" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(assembly_invasion_2<span class="sc">$</span>graph, assembly_invasion_2<span class="sc">$</span>info)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-56-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>And allowing four invasions in one go (as in top-down assembly):</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="ecological-assembly.html#cb50-1" tabindex="-1"></a>assembly_invasion_4 <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A, <span class="dv">4</span>)</span>
<span id="cb50-2"><a href="ecological-assembly.html#cb50-2" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(assembly_invasion_4<span class="sc">$</span>graph, assembly_invasion_4<span class="sc">$</span>info)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-57-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>For simplicity, let’s stick with the case in which only a single species enter in the local community at every invasion event.</p>
</div>
<div id="properties-of-the-assembly-graph" class="section level3 hasAnchor" number="7.4.5">
<h3><span class="header-section-number">7.4.5</span> Properties of the assembly graph<a href="ecological-assembly.html#properties-of-the-assembly-graph" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Accessibility</strong>: which states can we reach starting from the bare ground, and performing invasions of (say) one species at a time? We call states that can be built in this way “accessible”. Translated into graph properties, we call a state accessible if there is a path leading from the state 0 to the community of interest (“an assembly path”). If all states are accessible, we call the graph itself accessible. States that are not accessible can be reached by top-down, but not bottom-up assembly.</p>
<p><strong>Cycles</strong>: directed cycles in the graph translate into sub-communities for which invasions drive the system in a cyclic composition (a generalization of rock-paper-scissors!).</p>
<p><strong>Assembly endpoints</strong>: if a node has no outgoing edges, assembly will stop once the corresponding state has been reached. We call this state an “assembly endpoint”. A more complex type of assembly endpoint is that in which there is a cycle connecting two or more communities, and the cycle as a whole has no outgoing edges.</p>
<div class="float">
<img src="dat/assembly_cases.png" alt="Possible assembly graphs" />
<div class="figcaption">Possible assembly graphs</div>
</div>
</div>
<div id="assembly-graphs-for-glv" class="section level3 hasAnchor" number="7.4.6">
<h3><span class="header-section-number">7.4.6</span> Assembly graphs for GLV<a href="ecological-assembly.html#assembly-graphs-for-glv" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For GLV with symmetric, competitive interactions (actually, for a slightly more general case), <span class="citation">Serván and Allesina (<a href="#ref-servan2020tractable">2020</a>)</span> proved that:</p>
<ul>
<li><p>For a species pool of competitors (i.e., a given <span class="math inline">\(r &gt; 0\)</span> and a symmetric matrix <span class="math inline">\(A &lt; 0\)</span>), <strong>the bottom-up assembly endpoints are the same as the endpoints for top-down assembly</strong>.</p></li>
<li><p>Moreover, <strong>the assembly graph is accessible</strong>—therefore we can build any feasible, stable sub-community from the ground up. In fact, for each sub-community we can find the “shortest” assembly path, which we can construct without any extinction.</p></li>
<li><p><strong>The assembly graph is acyclic</strong>, meaning that we will never observe communities with cyclic compositions.</p></li>
<li><p><strong>Every walk on the assembly graph eventually reaches a sink</strong>, and, when <span class="math inline">\(A\)</span> is stable, <strong>the sink is unique</strong>. This means that for this type of species pool, historical contingencies (<span class="citation">Fukami (<a href="#ref-fukami2015historical">2015</a>)</span>) are impossible—if we wait for long enough, the system will always reach the same state, thereby erasing any trace of the assembly history.</p></li>
</ul>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="ecological-assembly.html#cb51-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5</span>)</span>
<span id="cb51-2"><a href="ecological-assembly.html#cb51-2" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_competitive_stable</span>(<span class="dv">6</span>)</span>
<span id="cb51-3"><a href="ecological-assembly.html#cb51-3" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">6</span>)</span>
<span id="cb51-4"><a href="ecological-assembly.html#cb51-4" tabindex="-1"></a>ag <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A)</span>
<span id="cb51-5"><a href="ecological-assembly.html#cb51-5" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(ag<span class="sc">$</span>graph, ag<span class="sc">$</span>info)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-58-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Community “29” is composed of species 1, 3, 4, 5. We can assemble this state without extinctions starting from the bare ground: first invade with species 4, sending the community to state “8”; then, add species 3, moving the state to “12”; then add species 5, moving to “28”, and finally add species 1, reaching “29”.</p>
<p>This fact has important consequences: for example, we can assemble the final state without the need for “transient invaders” (or “stepping stone species”, i.e., species that allow the assembly to proceed, but then disappear). Recently <span class="citation">Amor et al. (<a href="#ref-amor2020transient">2020</a>)</span> demonstrated experimentally the occurrence of transient invaders: they co-cultured <em>Corynebacterium ammoniagenes</em> (Ca) and <em>Lactobacillus plantarum</em> (Lp) and showed that one species displaces the other, with the identity of the winner depending on initial conditions (bistability). When the environment is dominated by Lp, Ca cannot invade. However, if first we introduce <em>Pseudomonas chlororaphis</em> (Pc), and then invade with Ca, Lp-dominated environments can be invaded by Ca—and Pc disappears without a trace. This is their Figure 1:</p>
<center>
<img src="https://www.science.org/cms/10.1126/sciadv.aay8676/asset/6521bd02-6546-403f-bfa8-5914e49e7dbb/assets/graphic/aay8676-f1.jpeg">
</center>
</div>
<div id="relationship-with-lyapunov-functions" class="section level3 hasAnchor" number="7.4.7">
<h3><span class="header-section-number">7.4.7</span> Relationship with Lyapunov functions<a href="ecological-assembly.html#relationship-with-lyapunov-functions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If a graph has the properties above (acyclic, single source, single sink), then there is a way to order the nodes such that all edges point in the same direction (topological sorting). In this case, we can devise an “energy” associated with each community state such that assembly “maximizes” the quantity, connecting “low-energy” states to “high-energy ones” via invasion.</p>
<p>In fact, we can even write down such a function—it is exactly the Lyapunov function devised by <span class="citation">MacArthur (<a href="#ref-macarthur1970species">1970</a>, Note: the first paper ever in Theoretical Population Biology!)</span> for GLV with symmetric, competitive interactions:</p>
<p><span class="math display">\[
V(x) = 2 \sum_i r_i x_i + \sum_{i,j} A_{ij} x_i x_j
\]</span></p>
<p>MacArthur proved that this quantity is maximized through the dynamics when the the equilibrium exists and <span class="math inline">\(A\)</span> is stable. At equilibrium, we find <span class="math inline">\(V(x^\star) = \sum_i r_i x_i^\star\)</span>. During assembly, community composition changes such that <span class="math inline">\(V(\bar{x})\)</span> is maximized through the assembly process. Note that when <span class="math inline">\(r_i = 1\)</span> for all species, then <span class="math inline">\(V(\bar{x})\)</span> is simply the total biomass in the community. As such, through invasions the total biomass in the community is maximized.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="ecological-assembly.html#cb52-1" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(ag<span class="sc">$</span>graph, ag<span class="sc">$</span>info, <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-60-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Now a case with equal growth rates:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="ecological-assembly.html#cb53-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10</span>)</span>
<span id="cb53-2"><a href="ecological-assembly.html#cb53-2" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_competitive_stable</span>(<span class="dv">5</span>)</span>
<span id="cb53-3"><a href="ecological-assembly.html#cb53-3" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">5</span>) <span class="co"># same growth rates!</span></span>
<span id="cb53-4"><a href="ecological-assembly.html#cb53-4" tabindex="-1"></a>ag2 <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph</span>(r, A)</span>
<span id="cb53-5"><a href="ecological-assembly.html#cb53-5" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(ag2<span class="sc">$</span>graph, ag2<span class="sc">$</span>info, <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-61-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>The community “15” is composed of species 1, 2, 3, and 4, while community “30” of species 2, 3, 4, and 5. Let’s compute their biomasses:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="ecological-assembly.html#cb54-1" tabindex="-1"></a><span class="co"># community 15</span></span>
<span id="cb54-2"><a href="ecological-assembly.html#cb54-2" tabindex="-1"></a>biom_15 <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">solve</span>(A[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="sc">-</span>r[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]))</span>
<span id="cb54-3"><a href="ecological-assembly.html#cb54-3" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;Community 15:&quot;</span>, biom_15))</span></code></pre></div>
<p>[1] “Community 15: 0.974135821684639”</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="ecological-assembly.html#cb55-1" tabindex="-1"></a><span class="co"># community 30</span></span>
<span id="cb55-2"><a href="ecological-assembly.html#cb55-2" tabindex="-1"></a>biom_30 <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">solve</span>(A[<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>], <span class="sc">-</span>r[<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>]))</span>
<span id="cb55-3"><a href="ecological-assembly.html#cb55-3" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;Community 30:&quot;</span>, biom_30))</span></code></pre></div>
<p>[1] “Community 30: 0.855081379845787”</p>
<p>This result would have pleased Cowles, Gleason, Odum, and many of the pioneers of succession! For these (restrictive) conditions, assembly is indeed an orderly, predictable process, culminating in a “climax” with a clear ecological interpretation. (For a similar result in a completely different context, see <span class="citation">Suweis et al. (<a href="#ref-suweis2013emergence">2013</a>)</span>).</p>
</div>
<div id="how-many-assembly-endpoints" class="section level3 hasAnchor" number="7.4.8">
<h3><span class="header-section-number">7.4.8</span> How many assembly endpoints?<a href="ecological-assembly.html#how-many-assembly-endpoints" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>What happens when the matrix <span class="math inline">\(A\)</span> is still nonpositive and symmetric, but not stable? In this case, it is Gleason who can laugh—for unstable matrices we can have several assembly endpoints, thereby reinstating the role of chance in determining the ultimate fate of the community.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="ecological-assembly.html#cb56-1" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/general_code_assembly.R&quot;</span>)</span>
<span id="cb56-2"><a href="ecological-assembly.html#cb56-2" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;dat/build_assembly_graph_unstable.R&quot;</span>)</span>
<span id="cb56-3"><a href="ecological-assembly.html#cb56-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5</span>)</span>
<span id="cb56-4"><a href="ecological-assembly.html#cb56-4" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb56-5"><a href="ecological-assembly.html#cb56-5" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">build_competitive_unstable</span>(n)</span>
<span id="cb56-6"><a href="ecological-assembly.html#cb56-6" tabindex="-1"></a>r <span class="ot">&lt;-</span> <span class="fu">runif</span>(n)</span>
<span id="cb56-7"><a href="ecological-assembly.html#cb56-7" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A)</span>
<span id="cb56-8"><a href="ecological-assembly.html#cb56-8" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(tmp<span class="sc">$</span>graph, tmp<span class="sc">$</span>info, <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-63-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Note that the Lyapunov function still holds (locally): when moving on the graph, the “energy” is always increased.</p>
<p>The number of “assembly endpoints” (i.e., saturated equilibria) depends on the stability of the matrix—what if we reduce the stability even further?</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="ecological-assembly.html#cb57-1" tabindex="-1"></a><span class="co"># these are the diagonal elements</span></span>
<span id="cb57-2"><a href="ecological-assembly.html#cb57-2" tabindex="-1"></a><span class="fu">diag</span>(A)</span></code></pre></div>
<p>[1] -1.1244467 -1.7740170 -0.9204892 -1.4486570 -0.1384223 -2.9264971
[7] -0.3754522 -2.0440457 -1.1836697 -0.2244013</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="ecological-assembly.html#cb58-1" tabindex="-1"></a><span class="co"># and these the eigenvalues</span></span>
<span id="cb58-2"><a href="ecological-assembly.html#cb58-2" tabindex="-1"></a><span class="fu">eigen</span>(A, <span class="at">symmetric =</span> <span class="cn">TRUE</span>, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span></code></pre></div>
<p>[1] 6.6159328 5.3603879 3.1108237 1.1950182 -0.1931400 -0.5697615
[7] -2.2641869 -3.2814435 -3.5994749 -18.5342540</p>
<p>Let’s add <span class="math inline">\(0.13\)</span> to the diagonal (such that the fifth species has almost no self-regulation)—this shifts all of the eigenvalues to the right:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="ecological-assembly.html#cb59-1" tabindex="-1"></a>A2 <span class="ot">&lt;-</span> A</span>
<span id="cb59-2"><a href="ecological-assembly.html#cb59-2" tabindex="-1"></a><span class="fu">diag</span>(A2) <span class="ot">&lt;-</span> <span class="fu">diag</span>(A2) <span class="sc">+</span> <span class="fl">0.13</span></span>
<span id="cb59-3"><a href="ecological-assembly.html#cb59-3" tabindex="-1"></a><span class="fu">eigen</span>(A2, <span class="at">symmetric =</span> <span class="cn">TRUE</span>, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span></code></pre></div>
<p>[1] 6.74593280 5.49038786 3.24082373 1.32501822 -0.06313999
[6] -0.43976150 -2.13418688 -3.15144351 -3.46947493 -18.40425400</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="ecological-assembly.html#cb60-1" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A2)</span>
<span id="cb60-2"><a href="ecological-assembly.html#cb60-2" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(tmp<span class="sc">$</span>graph, tmp<span class="sc">$</span>info, <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-66-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Similarly, by shifting the eigenvalues to the left, we can get access larger communities:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="ecological-assembly.html#cb61-1" tabindex="-1"></a>A3 <span class="ot">&lt;-</span> A</span>
<span id="cb61-2"><a href="ecological-assembly.html#cb61-2" tabindex="-1"></a><span class="fu">diag</span>(A3) <span class="ot">&lt;-</span> <span class="fu">diag</span>(A3) <span class="sc">-</span> <span class="dv">2</span></span>
<span id="cb61-3"><a href="ecological-assembly.html#cb61-3" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">build_assembly_graph_unstable</span>(r, A3)</span>
<span id="cb61-4"><a href="ecological-assembly.html#cb61-4" tabindex="-1"></a><span class="fu">plot_assembly_graph</span>(tmp<span class="sc">$</span>graph, tmp<span class="sc">$</span>info, <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-67-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Interestingly, how the number of assembly endpoints changes when we change the parameters of the model is an open problem—in fact we haven’t even characterized the worse-case scenario. See <span class="citation">Biroli et al. (<a href="#ref-biroli2018marginally">2018</a>)</span> for a derivation showing that they should be growing exponentially with size.</p>
</div>
<div id="build-your-own-assembly-graph" class="section level3 hasAnchor" number="7.4.9">
<h3><span class="header-section-number">7.4.9</span> Build your own assembly graph!<a href="ecological-assembly.html#build-your-own-assembly-graph" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Building the assembly graph is computationally very expensive, even with all these results at hand. Fortunately, we do not need to integrate the dynamics (a point that was greatly debated in the literature, see <span class="citation">Morton et al. (<a href="#ref-morton1996models">1996</a>)</span> for one of the rare cases in which a debate ends up with consensus between opposing factions).</p>
<p>Currently, the algorithm can be sketched as:</p>
<ul>
<li>for each of the <span class="math inline">\(2^n\)</span> possible sub-communities (ranging from bare ground to all species present), determine whether the sub-community is feasible and stable (this is the most computationally expensive step);</li>
<li>now go through all the feasible, stable communities; for each determine the “neighbor” communities that can be reached with (say, one) invasion(s).</li>
<li>call <span class="math inline">\(S\)</span> the feasible, stable community we are considering, and add species <span class="math inline">\(j\)</span> (or multiple species), and check that <span class="math inline">\(j\)</span> can invade when rare. If it cannot, move to the next neighbor; If it can, there are two cases:
<ul>
<li>if the community <span class="math inline">\(\{S, j\} = S&#39;\)</span> is feasible and stable, draw an edge <span class="math inline">\(S \to S&#39;\)</span></li>
<li>if the community <span class="math inline">\(\{S, j\} = S&#39;\)</span> is not feasible and stable, the system will collapse to a smaller sub-community; to determine which sub-community it will collapse to, <strong>check all the possible sub-communities</strong> of <span class="math inline">\(S&#39;\)</span>, and take the one with the greatest <span class="math inline">\(V(x^\star)\)</span>, <span class="math inline">\(S&#39;&#39;\)</span>. Draw an an edge <span class="math inline">\(S \to S&#39;&#39;\)</span></li>
</ul></li>
</ul>
<p>Can a better (faster, more efficient) algorithm be devised?</p>
<div class="boxhomework">
<p><strong>Homework 5</strong></p>
<p>Build the assembly graph for a GLV model in which the growth rates of all species are the same, and interactions are given by phylogenetic relatedness.</p>
<p>In particular, consider an ultrametric phylogenetic tree <span class="math inline">\(T\)</span>, and the “variance-covariance” matrix induced by the tree, <span class="math inline">\(\Sigma\)</span>. Because the tree is ultrametric, then <span class="math inline">\(\Sigma_{ii} = 1\)</span> for all <span class="math inline">\(i\)</span>. The off-diagonal elements <span class="math inline">\(\Sigma_{ij}\)</span> express the proportion of shared ancestry (and as such <span class="math inline">\(1\geq \Sigma_{ij} \geq 0\)</span>).</p>
<p>Take the GLV model:</p>
<p><span class="math display">\[
\dfrac{dx}{dt}=D(x)(1 - \Sigma x)
\]</span></p>
<p>Notice that <span class="math inline">\(\Sigma\)</span> is positive (semi-)definite by construction. Now build the assembly graph: does it have any interesting property?</p>
</div>
</div>
<div id="conclusions" class="section level3 hasAnchor" number="7.4.10">
<h3><span class="header-section-number">7.4.10</span> Conclusions<a href="ecological-assembly.html#conclusions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We have explored the problem of ecological assembly by working with the Generalized Lotka-Volterra model and three main assumptions: 1) invasions are rare, such that invaders always find the local community at the attractor; 2) invasion sizes are small; and 3) dynamics always lead to equilibria–thereby allowing us to determine the unique outcome of any invasion event.</p>
<p>These three assumptions made the study of assembly <strong>tractable, and yet not trivial</strong>, allowing us to map the complex dynamics of the system into an <strong>assembly graph</strong>: the properties of this graph translate directly into ecological properties of the assembly process.</p>
<p>For competitive Lotka-Volterra with symmetric interactions, when the interaction matrix is stable, all assembly paths will eventually lead to the same state (the “saturated equilibrium” we’ve seen in Lectures 1 and 2). In this case, the history that led to the community cannot be reconstructed from the final state. When the matrix of interactions is unstable, on the other hand, the system can end up in different places depending on the history of assembly.</p>
<p>We have also found that for these cases, <strong>top-down</strong> (all in one go) and <strong>bottom-up</strong> (sequential) assembly can reach exactly the same states. Top-down assembly experiments (e.g., with bacterial communities) are much easier to perform than bottom-up ones, and I believe that this approach has been underexploited (both theoretically, and experimentally).</p>
<p>Much more work remains to be done to reach a general theory of ecological assembly (even for GLV), but the results shown here provide a good springboard, and an expectation, for the study of more complex cases.</p>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-amor2020transient" class="csl-entry">
Amor, D. R., C. Ratzke, and J. Gore. 2020. Transient invaders can induce shifts between alternative stable states of microbial communities. Science advances 6:eaay8676.
</div>
<div id="ref-biroli2018marginally" class="csl-entry">
Biroli, G., G. Bunin, and C. Cammarota. 2018. Marginally stable equilibria in critical ecosystems. New Journal of Physics 20:083051.
</div>
<div id="ref-bittleston2020context" class="csl-entry">
Bittleston, L. S., M. Gralka, G. E. Leventhal, I. Mizrahi, and O. X. Cordero. 2020. Context-dependent dynamics lead to the assembly of functionally distinct microbial communities. Nature communications 11:1–10.
</div>
<div id="ref-cadotte2016phylogenies" class="csl-entry">
Cadotte, M. W., and T. J. Davies. 2016. Phylogenies in ecology: A guide to concepts and methods. Princeton University Press.
</div>
<div id="ref-capitan2009statistical" class="csl-entry">
Capitán, J. A., J. A. Cuesta, and J. Bascompte. 2009. Statistical mechanics of ecosystem assembly. Physical Review Letters 103:168101.
</div>
<div id="ref-diamond1975assembly" class="csl-entry">
Diamond, J. M. 1975. Assembly of species communities. Ecology and evolution of communities:342–444.
</div>
<div id="ref-drake1991community" class="csl-entry">
Drake, J. A. 1991. Community-assembly mechanics and the structure of an experimental species ensemble. The American Naturalist 137:1–26.
</div>
<div id="ref-fukami2015historical" class="csl-entry">
Fukami, T. 2015. Historical contingency in community assembly: Integrating niches, species pools, and priority effects. Annual Review of Ecology, Evolution, and Systematics 46:1–23.
</div>
<div id="ref-gotelli1999communities" class="csl-entry">
Gotelli, N. J. 1999. How do communities come together? Science 286:1684–1685.
</div>
<div id="ref-gotelli1996null" class="csl-entry">
Gotelli, N. J., and G. R. Graves. 1996. Null models in ecology.
</div>
<div id="ref-gould1979spandrels" class="csl-entry">
Gould, S. J., and R. C. Lewontin. 1979. The spandrels of san marco and the panglossian paradigm: A critique of the adaptationist programme. Proceedings of the Royal Society of London. Series B. Biological Sciences 205:581–598.
</div>
<div id="ref-hang1993assembly" class="csl-entry">
Hang-Kwang, L., and S. L. Pimm. 1993. The assembly of ecological communities: A minimalist approach. Journal of Animal Ecology:749–765.
</div>
<div id="ref-hubbell2001unified" class="csl-entry">
Hubbell, S. P. 2001. The unified neutral theory of biodiversity and biogeography. Princeton University Press.
</div>
<div id="ref-law1993alternative" class="csl-entry">
Law, R., and R. D. Morton. 1993. Alternative permanent states of ecological communities. Ecology 74:1347–1361.
</div>
<div id="ref-law1996permanence" class="csl-entry">
Law, R., and R. D. Morton. 1996. Permanence and the assembly of ecological communities. Ecology 77:762–775.
</div>
<div id="ref-macarthur1970species" class="csl-entry">
MacArthur, R. 1970. Species packing and competitive equilibrium for many species. Theoretical population biology 1:1–11.
</div>
<div id="ref-maynard2018network" class="csl-entry">
Maynard, D. S., C. A. Serván, and S. Allesina. 2018. Network spandrels reflect ecological assembly. Ecology letters 21:324–334.
</div>
<div id="ref-morton1996models" class="csl-entry">
Morton, R. D., R. Law, S. L. Pimm, and J. A. Drake. 1996. On models for assembling ecological communities. Oikos:493–499.
</div>
<div id="ref-pielou1977mathematical" class="csl-entry">
Pielou, E. C. 1977. Mathematical ecology. Wiley,.
</div>
<div id="ref-schreiber2004simple" class="csl-entry">
Schreiber, S. J., and S. Rittenhouse. 2004. From simple rules to cycling in community assembly. Oikos 105:349–358.
</div>
<div id="ref-servan2020tractable" class="csl-entry">
Serván, C. A., and S. Allesina. 2020. Tractable models of ecological assembly. bioRxiv.
</div>
<div id="ref-servan2018coexistence" class="csl-entry">
Serván, C. A., J. A. Capitán, J. Grilli, K. E. Morrison, and S. Allesina. 2018. Coexistence of many species in random ecosystems. Nature ecology &amp; evolution 2:1237.
</div>
<div id="ref-sigmuiud1995darwin" class="csl-entry">
Sigmund, K. 1995. Darwin’s "circles of complexity": Assembling ecological communities. Complexity 1:40–44.
</div>
<div id="ref-sole2006network" class="csl-entry">
Solé, R. V., and S. Valverde. 2006. Are network motifs the spandrels of cellular complexity? Trends in ecology &amp; evolution 21:419–422.
</div>
<div id="ref-suweis2013emergence" class="csl-entry">
Suweis, S., F. Simini, J. R. Banavar, and A. Maritan. 2013. Emergence of structural and dynamical properties of ecological mutualistic networks. Nature 500:449–452.
</div>
<div id="ref-weiher2001ecological" class="csl-entry">
Weiher, E., and P. Keddy. 2001. Ecological assembly rules: Perspectives, advances, retreats. Cambridge University Press.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="game-theory-and-replicator-dynamics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="metapopulation-dynamics-on-networks.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": null,
    "text": null
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["Theoretical_Community_Ecology.pdf", "Theoretical_Community_Ecology.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
