<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lecture 5 Predicting coexistence in ecological communities | Theoretical Community Ecology</title>
  <meta name="description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Lecture 5 Predicting coexistence in ecological communities | Theoretical Community Ecology" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lecture 5 Predicting coexistence in ecological communities | Theoretical Community Ecology" />
  
  <meta name="twitter:description" content="A graduate course in theoretical community ecology focusing on systems with multiple species." />
  

<meta name="author" content="Stefano Allesina" />


<meta name="date" content="2025-04-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="stability-of-large-ecological-communities.html"/>
<link rel="next" href="game-theory-and-replicator-dynamics.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Theoretical Community Ecology</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preliminaries</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#approach-and-target-audience"><i class="fa fa-check"></i>Approach and target audience</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#topics"><i class="fa fa-check"></i>Topics</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#notation"><i class="fa fa-check"></i>Notation</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#grading-and-code-of-conduct"><i class="fa fa-check"></i>Grading and code of conduct</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#homework-80"><i class="fa fa-check"></i>Homework (80%)</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#review-20"><i class="fa fa-check"></i>Review (20%)</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#accommodation-for-students-with-disabilities"><i class="fa fa-check"></i>Accommodation for students with disabilities</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#a-note-about-biographies"><i class="fa fa-check"></i>A note about biographies</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#sources"><i class="fa fa-check"></i>Sources</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html"><i class="fa fa-check"></i><b>1</b> Models for a single population</a>
<ul>
<li class="chapter" data-level="1.1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#types-of-dynamical-systems"><i class="fa fa-check"></i><b>1.1</b> Types of dynamical systems</a></li>
<li class="chapter" data-level="1.2" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#initial-value-problems"><i class="fa fa-check"></i><b>1.2</b> Initial-value problems</a></li>
<li class="chapter" data-level="1.3" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#solvable-model-exponential-growth"><i class="fa fa-check"></i><b>1.3</b> Solvable model: Exponential growth</a></li>
<li class="chapter" data-level="1.4" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#solvable-model-logistic-growth"><i class="fa fa-check"></i><b>1.4</b> Solvable model: Logistic growth</a></li>
<li class="chapter" data-level="1.5" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#qualitative-analysis-of-models-for-a-single-population"><i class="fa fa-check"></i><b>1.5</b> Qualitative analysis of models for a single population</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#stability-using-derivatives"><i class="fa fa-check"></i><b>1.5.1</b> Stability using derivatives</a></li>
<li class="chapter" data-level="1.5.2" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#what-are-we-doing-when-probing-stability-using-derivatives"><i class="fa fa-check"></i><b>1.5.2</b> What are we doing when probing stability using derivatives?</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#bifurcations"><i class="fa fa-check"></i><b>1.6</b> Bifurcations</a></li>
<li class="chapter" data-level="1.7" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#long-term-behavior-of-1-d-systems"><i class="fa fa-check"></i><b>1.7</b> Long term behavior of 1-D systems</a></li>
<li class="chapter" data-level="1.8" data-path="models-for-a-single-population.html"><a href="models-for-a-single-population.html#lyapunov-functions"><i class="fa fa-check"></i><b>1.8</b> Lyapunov functions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html"><i class="fa fa-check"></i><b>2</b> Models for two populations</a>
<ul>
<li class="chapter" data-level="2.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#qualitative-analysis-of-models-for-two-populations"><i class="fa fa-check"></i><b>2.1</b> Qualitative analysis of models for two populations</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#isoclines-of-null-growth"><i class="fa fa-check"></i><b>2.1.1</b> Isoclines of null growth</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#local-stability-analysis"><i class="fa fa-check"></i><b>2.2</b> Local stability analysis</a></li>
<li class="chapter" data-level="2.3" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#stability-analysis-of-the-lotka-volterra-predator-prey-model"><i class="fa fa-check"></i><b>2.3</b> Stability analysis of the Lotka-Volterra Predator-Prey model</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#constant-of-motion-for-lotka-volterra-predator-prey"><i class="fa fa-check"></i><b>2.3.1</b> Constant of motion for Lotka-Volterra Predator-Prey</a></li>
<li class="chapter" data-level="2.3.2" data-path="models-for-two-populations.html"><a href="models-for-two-populations.html#analysis-of-predator-prey-model-with-logistic-growth-for-the-prey"><i class="fa fa-check"></i><b>2.3.2</b> Analysis of Predator-Prey model with logistic growth for the prey</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html"><i class="fa fa-check"></i><b>3</b> Generalized Lotka-Volterra model</a>
<ul>
<li class="chapter" data-level="3.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#formulation"><i class="fa fa-check"></i><b>3.1</b> Formulation</a></li>
<li class="chapter" data-level="3.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#a-single-population"><i class="fa fa-check"></i><b>3.2</b> A single population</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#metapopulation-dynamics"><i class="fa fa-check"></i><b>3.2.1</b> Metapopulation dynamics</a></li>
<li class="chapter" data-level="3.2.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#s-i-s-model"><i class="fa fa-check"></i><b>3.2.2</b> S-I-S model</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#multi-species-dynamics"><i class="fa fa-check"></i><b>3.3</b> Multi-species dynamics</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#existence-of-an-equilibrium"><i class="fa fa-check"></i><b>3.3.1</b> Existence of an equilibrium</a></li>
<li class="chapter" data-level="3.3.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#types-of-dynamics"><i class="fa fa-check"></i><b>3.3.2</b> Types of dynamics</a></li>
<li class="chapter" data-level="3.3.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#equilibrium-as-time-average"><i class="fa fa-check"></i><b>3.3.3</b> Equilibrium as time-average</a></li>
<li class="chapter" data-level="3.3.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#coexistence-requires-a-feasible-equilibrium"><i class="fa fa-check"></i><b>3.3.4</b> Coexistence requires a feasible equilibrium</a></li>
<li class="chapter" data-level="3.3.5" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#local-stability-of-the-coexistence-equilibrium"><i class="fa fa-check"></i><b>3.3.5</b> Local stability of the coexistence equilibrium</a></li>
<li class="chapter" data-level="3.3.6" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#lyapunov-diagonal-stability-and-global-stability"><i class="fa fa-check"></i><b>3.3.6</b> Lyapunov diagonal stability and global stability</a></li>
<li class="chapter" data-level="3.3.7" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#saturated-equilibria"><i class="fa fa-check"></i><b>3.3.7</b> Saturated equilibria</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#macarthurs-consumer-resource-model"><i class="fa fa-check"></i><b>3.4</b> MacArthur’s consumer-resource model</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#existence-of-an-equilibrium-1"><i class="fa fa-check"></i><b>3.4.1</b> Existence of an equilibrium</a></li>
<li class="chapter" data-level="3.4.2" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#global-stability-1"><i class="fa fa-check"></i><b>3.4.2</b> Global stability</a></li>
<li class="chapter" data-level="3.4.3" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#separation-of-time-scales"><i class="fa fa-check"></i><b>3.4.3</b> Separation of time-scales</a></li>
<li class="chapter" data-level="3.4.4" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#lyapunov-function-for-symmetric-lotka-volterra"><i class="fa fa-check"></i><b>3.4.4</b> Lyapunov function for symmetric Lotka-Volterra</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="generalized-lotka-volterra-model.html"><a href="generalized-lotka-volterra-model.html#further-readings"><i class="fa fa-check"></i><b>3.5</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html"><i class="fa fa-check"></i><b>4</b> Stability of large ecological communities</a>
<ul>
<li class="chapter" data-level="4.1" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#complexity-and-stability"><i class="fa fa-check"></i><b>4.1</b> Complexity and stability</a></li>
<li class="chapter" data-level="4.2" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#the-stability-of-random-ecosystems"><i class="fa fa-check"></i><b>4.2</b> The stability of random ecosystems</a></li>
<li class="chapter" data-level="4.3" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#accounting-for-interaction-types"><i class="fa fa-check"></i><b>4.3</b> Accounting for interaction types</a></li>
<li class="chapter" data-level="4.4" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#symmetric-matrices"><i class="fa fa-check"></i><b>4.4</b> Symmetric matrices</a></li>
<li class="chapter" data-level="4.5" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#covariance-matrices"><i class="fa fa-check"></i><b>4.5</b> Covariance matrices</a></li>
<li class="chapter" data-level="4.6" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#small-rank-perturbations"><i class="fa fa-check"></i><b>4.6</b> Small-rank perturbations</a></li>
<li class="chapter" data-level="4.7" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#structured-matrices"><i class="fa fa-check"></i><b>4.7</b> Structured matrices</a></li>
<li class="chapter" data-level="4.8" data-path="stability-of-large-ecological-communities.html"><a href="stability-of-large-ecological-communities.html#further-readings-1"><i class="fa fa-check"></i><b>4.8</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html"><i class="fa fa-check"></i><b>5</b> Predicting coexistence in ecological communities</a>
<ul>
<li class="chapter" data-level="5.1" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#background-diversity-and-ecosystem-functioning"><i class="fa fa-check"></i><b>5.1</b> Background: Diversity and Ecosystem Functioning</a></li>
<li class="chapter" data-level="5.2" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#example-data"><i class="fa fa-check"></i><b>5.2</b> Example data</a></li>
<li class="chapter" data-level="5.3" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#a-simple-statistical-framework"><i class="fa fa-check"></i><b>5.3</b> A simple statistical framework</a></li>
<li class="chapter" data-level="5.4" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#the-structure-of-equilibria-in-glv"><i class="fa fa-check"></i><b>5.4</b> The structure of equilibria in GLV</a></li>
<li class="chapter" data-level="5.5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#predicting-coexistence"><i class="fa fa-check"></i><b>5.5</b> Predicting coexistence</a></li>
<li class="chapter" data-level="5.6" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#scoring-matrices-b"><i class="fa fa-check"></i><b>5.6</b> Scoring matrices <span class="math inline">\(B\)</span></a></li>
<li class="chapter" data-level="5.7" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#example"><i class="fa fa-check"></i><b>5.7</b> Example</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#sum-of-squares"><i class="fa fa-check"></i><b>5.7.1</b> Sum of squares</a></li>
<li class="chapter" data-level="5.7.2" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#weighted-leas-squares"><i class="fa fa-check"></i><b>5.7.2</b> Weighted leas squares</a></li>
<li class="chapter" data-level="5.7.3" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#maximum-likelihood"><i class="fa fa-check"></i><b>5.7.3</b> Maximum likelihood</a></li>
<li class="chapter" data-level="5.7.4" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#leave-one-out-cross-validation"><i class="fa fa-check"></i><b>5.7.4</b> Leave-one-out cross validation</a></li>
<li class="chapter" data-level="5.7.5" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#simplified-models"><i class="fa fa-check"></i><b>5.7.5</b> Simplified models</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="predicting-coexistence-in-ecological-communities.html"><a href="predicting-coexistence-in-ecological-communities.html#further-readings-2"><i class="fa fa-check"></i><b>5.8</b> Further readings</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html"><i class="fa fa-check"></i><b>6</b> Game theory and replicator dynamics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#game-theory"><i class="fa fa-check"></i><b>6.1</b> Game theory</a></li>
<li class="chapter" data-level="6.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#bimatrix-games"><i class="fa fa-check"></i><b>6.2</b> Bimatrix games</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#example-the-prisoners-dilemma"><i class="fa fa-check"></i><b>6.2.1</b> Example: the prisoner’s dilemma</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#mixed-strategies"><i class="fa fa-check"></i><b>6.3</b> Mixed strategies</a></li>
<li class="chapter" data-level="6.4" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#nash-equilibria"><i class="fa fa-check"></i><b>6.4</b> Nash Equilibria</a></li>
<li class="chapter" data-level="6.5" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#symmetric-bimatrix-games"><i class="fa fa-check"></i><b>6.5</b> Symmetric bimatrix games</a></li>
<li class="chapter" data-level="6.6" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#evolutionary-stable-strategies"><i class="fa fa-check"></i><b>6.6</b> Evolutionary stable strategies</a></li>
<li class="chapter" data-level="6.7" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#replicator-dynamics"><i class="fa fa-check"></i><b>6.7</b> Replicator dynamics</a>
<ul>
<li class="chapter" data-level="6.7.1" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#nash-equilibria-are-equilibria-of-the-re"><i class="fa fa-check"></i><b>6.7.1</b> Nash equilibria are equilibria of the RE</a></li>
<li class="chapter" data-level="6.7.2" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#ess-are-globally-stable-equilibria-of-the-re"><i class="fa fa-check"></i><b>6.7.2</b> ESS are globally stable equilibria of the RE</a></li>
<li class="chapter" data-level="6.7.3" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#stability-of-the-re-with-symmetric-a"><i class="fa fa-check"></i><b>6.7.3</b> Stability of the RE with symmetric <span class="math inline">\(A\)</span></a></li>
<li class="chapter" data-level="6.7.4" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#invariants"><i class="fa fa-check"></i><b>6.7.4</b> Invariants</a></li>
<li class="chapter" data-level="6.7.5" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#rock-paper-scissor"><i class="fa fa-check"></i><b>6.7.5</b> Rock-paper-scissor</a></li>
<li class="chapter" data-level="6.7.6" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#equivalence-with-glv"><i class="fa fa-check"></i><b>6.7.6</b> Equivalence with GLV</a></li>
<li class="chapter" data-level="6.7.7" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#hypertournament-games"><i class="fa fa-check"></i><b>6.7.7</b> Hypertournament games</a></li>
<li class="chapter" data-level="6.7.8" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#lyapunov-function"><i class="fa fa-check"></i><b>6.7.8</b> Lyapunov function</a></li>
<li class="chapter" data-level="6.7.9" data-path="game-theory-and-replicator-dynamics.html"><a href="game-theory-and-replicator-dynamics.html#higher-order-interactions"><i class="fa fa-check"></i><b>6.7.9</b> Higher-order interactions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ecological-assembly.html"><a href="ecological-assembly.html"><i class="fa fa-check"></i><b>7</b> Ecological assembly</a>
<ul>
<li class="chapter" data-level="7.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#assembly"><i class="fa fa-check"></i><b>7.1</b> Assembly</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#what-makes-the-study-of-assembly-difficult"><i class="fa fa-check"></i><b>7.1.1</b> What makes the study of assembly difficult?</a></li>
<li class="chapter" data-level="7.1.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#ecological-assembly-without-tears"><i class="fa fa-check"></i><b>7.1.2</b> Ecological assembly without tears</a></li>
<li class="chapter" data-level="7.1.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#top-down-and-bottom-up-assembly"><i class="fa fa-check"></i><b>7.1.3</b> Top-down and bottom-up assembly</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#modeling-invasions"><i class="fa fa-check"></i><b>7.2</b> Modeling invasions</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#invasions-in-multispecies-communities"><i class="fa fa-check"></i><b>7.2.1</b> Invasions in multispecies communities</a></li>
<li class="chapter" data-level="7.2.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#lyapunov-stability-and-saturated-equilibria"><i class="fa fa-check"></i><b>7.2.2</b> Lyapunov stability and saturated equilibria</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#top-down-assembly"><i class="fa fa-check"></i><b>7.3</b> Top-down assembly</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#a-trivial-case"><i class="fa fa-check"></i><b>7.3.1</b> A trivial case</a></li>
<li class="chapter" data-level="7.3.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#simulating-final-composition"><i class="fa fa-check"></i><b>7.3.2</b> Simulating final composition</a></li>
<li class="chapter" data-level="7.3.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#a-random-zoo"><i class="fa fa-check"></i><b>7.3.3</b> A random zoo</a></li>
<li class="chapter" data-level="7.3.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#a-random-zoo-in-the-wild"><i class="fa fa-check"></i><b>7.3.4</b> A random zoo in the wild</a></li>
<li class="chapter" data-level="7.3.5" data-path="ecological-assembly.html"><a href="ecological-assembly.html#assembly-and-saturated-equilibrium"><i class="fa fa-check"></i><b>7.3.5</b> Assembly and saturated equilibrium</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#network-spandrels"><i class="fa fa-check"></i><b>7.4</b> Network spandrels</a></li>
<li class="chapter" data-level="7.5" data-path="ecological-assembly.html"><a href="ecological-assembly.html#bottom-up-assembly"><i class="fa fa-check"></i><b>7.5</b> Bottom-up assembly</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="ecological-assembly.html"><a href="ecological-assembly.html#an-assembly-graph"><i class="fa fa-check"></i><b>7.5.1</b> An assembly graph</a></li>
<li class="chapter" data-level="7.5.2" data-path="ecological-assembly.html"><a href="ecological-assembly.html#how-many-invasions"><i class="fa fa-check"></i><b>7.5.2</b> How many invasions?</a></li>
<li class="chapter" data-level="7.5.3" data-path="ecological-assembly.html"><a href="ecological-assembly.html#properties-of-the-assembly-graph"><i class="fa fa-check"></i><b>7.5.3</b> Properties of the assembly graph</a></li>
<li class="chapter" data-level="7.5.4" data-path="ecological-assembly.html"><a href="ecological-assembly.html#assembly-graphs-for-glv"><i class="fa fa-check"></i><b>7.5.4</b> Assembly graphs for GLV</a></li>
<li class="chapter" data-level="7.5.5" data-path="ecological-assembly.html"><a href="ecological-assembly.html#relationship-with-lyapunov-functions"><i class="fa fa-check"></i><b>7.5.5</b> Relationship with Lyapunov functions</a></li>
<li class="chapter" data-level="7.5.6" data-path="ecological-assembly.html"><a href="ecological-assembly.html#how-many-assembly-endpoints"><i class="fa fa-check"></i><b>7.5.6</b> How many assembly endpoints?</a></li>
<li class="chapter" data-level="7.5.7" data-path="ecological-assembly.html"><a href="ecological-assembly.html#build-your-own-assembly-graph"><i class="fa fa-check"></i><b>7.5.7</b> Build your own assembly graph!</a></li>
<li class="chapter" data-level="7.5.8" data-path="ecological-assembly.html"><a href="ecological-assembly.html#conclusions"><i class="fa fa-check"></i><b>7.5.8</b> Conclusions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html"><i class="fa fa-check"></i><b>8</b> Metapopulation dynamics on networks</a>
<ul>
<li class="chapter" data-level="8.1" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#metapopulations"><i class="fa fa-check"></i><b>8.1</b> Metapopulations</a></li>
<li class="chapter" data-level="8.2" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#hanski-ovaskainen-2000"><i class="fa fa-check"></i><b>8.2</b> Hanski Ovaskainen 2000</a></li>
<li class="chapter" data-level="8.3" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#sis-on-a-contact-network"><i class="fa fa-check"></i><b>8.3</b> SIS on a contact network</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#the-full-model"><i class="fa fa-check"></i><b>8.3.1</b> The full model</a></li>
<li class="chapter" data-level="8.3.2" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#the-mean-field-approximation"><i class="fa fa-check"></i><b>8.3.2</b> The mean-field approximation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="metapopulation-dynamics-on-networks.html"><a href="metapopulation-dynamics-on-networks.html#metapopulations-on-randomly-distributed-patches"><i class="fa fa-check"></i><b>8.4</b> Metapopulations on randomly distributed patches</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html"><i class="fa fa-check"></i><b>9</b> Quasi-polynomial systems</a>
<ul>
<li class="chapter" data-level="9.1" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#global-stability-2"><i class="fa fa-check"></i><b>9.1</b> Global stability</a></li>
<li class="chapter" data-level="9.2" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#from-qp-to-glv"><i class="fa fa-check"></i><b>9.2</b> From QP to GLV</a></li>
<li class="chapter" data-level="9.3" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#stability-of-glv"><i class="fa fa-check"></i><b>9.3</b> Stability of GLV</a></li>
<li class="chapter" data-level="9.4" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#stability-for-glv-embeddings"><i class="fa fa-check"></i><b>9.4</b> Stability for GLV embeddings</a></li>
<li class="chapter" data-level="9.5" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#numerical-analysis"><i class="fa fa-check"></i><b>9.5</b> Numerical analysis</a></li>
<li class="chapter" data-level="9.6" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#systems-with-many-equations"><i class="fa fa-check"></i><b>9.6</b> Systems with many equations</a></li>
<li class="chapter" data-level="9.7" data-path="quasi-polynomial-systems.html"><a href="quasi-polynomial-systems.html#glv-embedding-for-non-qp-systems"><i class="fa fa-check"></i><b>9.7</b> GLV embedding for non-QP systems</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Theoretical Community Ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="predicting-coexistence-in-ecological-communities" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Lecture 5</span> Predicting coexistence in ecological communities<a href="predicting-coexistence-in-ecological-communities.html#predicting-coexistence-in-ecological-communities" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><strong>Lesson plan</strong>:</p>
<ol style="list-style-type: decimal">
<li>We consider the problem of co-culturing several species taken from a pool in all possible combinations.</li>
<li>This type of experiments has been conducted to test hypotheses on the relationship between diversity and ecosystem functioning.</li>
<li>The number of combinations grows quickly with the size of the pool, making experiments difficult.</li>
<li>The difficulty is compounded by the fact that not all combinations are expected to lead to coexistence.</li>
<li>We parameterize a simple statistical model, and draw a connection with GLV dynamics.</li>
<li>We test whether our model is able to predict the outcomes of experiments out-of-fit.</li>
</ol>
<div id="background-diversity-and-ecosystem-functioning" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Background: Diversity and Ecosystem Functioning<a href="predicting-coexistence-in-ecological-communities.html#background-diversity-and-ecosystem-functioning" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="boxpaper">
<p><strong>Key paper: <span class="citation">Tilman et al. (<a href="#ref-tilman2001diversity">2001</a>)</span></strong></p>
<p>This paper present the results of the biodiversity-ecosystem functioning experiments carried out at Cedar Creek, Minnesota, by Dave Tilman and collaborators. These multi-year experiments are among the largest ever carried out in ecology, and have been hugely influential. The same type of experiments have been performed with different organisms, in the lab, and in mesocosms.</p>
</div>
<p>Ecologists have performed large experiments in which different assemblages of species are co-cultured. These experiments have been conducted with plants (for example, Biodiversity Ecosystem Functioning experiments e.g., <span class="citation">Hector et al. (<a href="#ref-hector1999plant">1999</a>)</span> <span class="citation">Tilman et al. (<a href="#ref-tilman2001diversity">2001</a>)</span>, <span class="citation">Cadotte (<a href="#ref-cadotte2013experimental">2013</a>)</span>) and in the laboratory using protozoan, algae or bacteria. Two commonly-encountered problems in this type of experiments have to to with the scaling of the number of experiments with the number of species, and with the probability of coexistence.</p>
<p><strong>Scale:</strong> How many communities can we form from a pool of <span class="math inline">\(n\)</span> species? We can culture a single species in isolation (<span class="math inline">\(n\)</span> possibilities), two species in pair (<span class="math inline">\(n(n-1) / 2\)</span> possibilities), and so on. The total is therefore:</p>
<p><span class="math display">\[
\sum_{j=1}^n \binom{n}{j} = 2^n -1
\]</span></p>
<p>And this is only considering the presence/absence of each species! Moreover, we might want to vary the initial conditions (e.g., starting two species at low/high abundance, equal abundance, high/low abundance), etc. Clearly, this makes trying all possible combinations unfeasible when <span class="math inline">\(n\)</span> is large enough. For example, for 10 species we can form 1023 assemblages, while with 20 more than a million!</p>
<p><strong>Coexistence:</strong> even if we could try all possible experiments, many assemblages would collapse to smaller communities because of extinctions. For example, pairs could become monocultures, triplets become pairs or monocultures, etc. As such, even if we were to try all possible combinations, we would end up observing a smaller set of “final communities”.</p>
<p>To guide experimentation, we need a way to be able to predict the (probable) outcome of experiments without having to run them all. Here we attempt to do so by examining a handful of experimental results, and using these data to parametrize a statistical model. The model provides a way to navigate the enormous space of possibilities, thereby suggesting “good” experiments that yield a large probability of coexistence.</p>
</div>
<div id="example-data" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Example data<a href="predicting-coexistence-in-ecological-communities.html#example-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For our explorations, we are going to use recent data from Ishizawa and colleagues, which you can find here:</p>
<p>H. Ishizawa, Y. Tashiro, D. Inoue, M. Ike, H. Futamata.
<a href="https://doi.org/10.1073/pnas.2312396121"><em>Learning beyond-pairwise interactions enables the bottom–up prediction of microbial community structure</em></a>
PNAS 121 (7) e2312396121 (2024).</p>
<p>The Authors inoculated duckweed (<em>Lemna minor</em>) with synthetic bacterial communities formed by all possible combinations of seven strains. To this end, they cultured the infected duckweed in flasks for 10 days. At the end of the experiment, they plated the communities on agar plates containing antibiotics that would allow the growth only of a particular strain. In this way, they were able to measure the final density of each of the seven strains in each of the <span class="math inline">\(2^7 - 1 = 127\)</span> possible communities, and conducted each experiment in replicate. The full data set reports the outcome of 692 separate experiments!</p>
<p>More modestly, here we are going to focus on a smaller pool of three strains taken from the seven available. We therefore have <span class="math inline">\(7\)</span> possible communities, ranging from a single strain growing in isolation to the three strains growing together. For example, a few of the measurements are:</p>
<table>
<thead>
<tr class="header">
<th>DW067</th>
<th>DW102</th>
<th>DW145</th>
<th>community</th>
<th>replicate</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td></td>
<td></td>
<td>0.8</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>0.812</td>
<td>1</td>
<td>2</td>
</tr>
<tr class="odd">
<td></td>
<td>0.529</td>
<td></td>
<td>2</td>
<td>1</td>
</tr>
<tr class="even">
<td></td>
<td>0.356</td>
<td></td>
<td>2</td>
<td>2</td>
</tr>
<tr class="odd">
<td></td>
<td>0.703</td>
<td>0.909</td>
<td>3</td>
<td>1</td>
</tr>
<tr class="even">
<td></td>
<td>0.395</td>
<td>0.661</td>
<td>3</td>
<td>2</td>
</tr>
<tr class="odd">
<td>0.386</td>
<td>0.602</td>
<td>0.563</td>
<td>4</td>
<td>1</td>
</tr>
<tr class="even">
<td>0.379</td>
<td>0.774</td>
<td>0.293</td>
<td>4</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>We can therefore associate each measurement with a) the strain being measured, <span class="math inline">\(i\)</span>; b) the community in which <span class="math inline">\(i\)</span> was grown, <span class="math inline">\(k\)</span>; and c) the (biological) replicate experiment, <span class="math inline">\(r\)</span>.</p>
</div>
<div id="a-simple-statistical-framework" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> A simple statistical framework<a href="predicting-coexistence-in-ecological-communities.html#a-simple-statistical-framework" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The simplest model we can write for this type of data is one in which the outcomes of replicate experiments are independent samples from a distribution:</p>
<p><span class="math display">\[
\tilde{x}^{(k,r)}_i \sim Q_i\left(x^{(k)}_i, \gamma_i^{(k)}\right)
\]</span></p>
<p>where <span class="math inline">\(\tilde{x}_i^{(k, r)}\)</span> is the <em>observed</em> density of population <span class="math inline">\(i\)</span> for the <span class="math inline">\(r\)</span> replicate in which population <span class="math inline">\(i\)</span> is grown in community <span class="math inline">\(k\)</span>. The value <span class="math inline">\(x_i^{(k)}\)</span> represents the <em>true</em> mean of the distribution (i.e., the average if we were to conduct many replicates—hence it does not depend on <span class="math inline">\(r\)</span>), and <span class="math inline">\(\gamma_i^{(k)}\)</span> is a parameter (or several parameters) controlling the shape of the distribution <span class="math inline">\(Q_i\)</span>.</p>
<p>Implicitly, we are making a very strong assumption: if we observe the community <span class="math inline">\(k\)</span>, it is always found around <span class="math inline">\(x^{(k)}\)</span>—i.e., we cannot have <em>true multistability</em>, in which, depending on initial conditions, we end up with different outcomes in which all populations are present (we can still have that, depending on initial condition, the system ends up at different points/attractors, as long as they have different compositions).</p>
<p>This model requires estimating all the <span class="math inline">\(x_i^{(k)}\)</span> and <span class="math inline">\(\gamma_i^{(k)}\)</span>, and is therefore not very useful. To make the model applicable to real data, we make another strong assumption:</p>
<p><span class="math display">\[
x_i^{(k)} = \alpha_i - \sum_{j \in k;\, j\neq i } \beta_{ij} x_j^{(k)}
\]</span></p>
<p>The interpretation is simple: if population <span class="math inline">\(i\)</span> is grown by itself, it will reach the carrying capacity <span class="math inline">\(\alpha_i\)</span>; if other species are co-cultured along with <span class="math inline">\(i\)</span>, they will change the final density of <span class="math inline">\(i\)</span> according to their density (<span class="math inline">\(x_j^{(k)}\)</span>) and an interaction term <span class="math inline">\(\beta_{ij}\)</span>.</p>
<p>We perform some manipulations:</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{j \in k;\, j\neq i } \beta_{ij} x_j^{(k)} + x_i^{(k)} &amp;= \alpha_i \\
\sum_{j \in k;\, j\neq i } \frac{\beta_{ij}}{\alpha_i} x_j^{(k)} + \frac{1}{\alpha_i} x_i^{(k)} &amp;= 1 \\
\sum_{j \in k;\, j\neq i } B_{ij} x_j^{(k)} + B_{ii} x_i^{(k)} &amp;= 1 \\
\sum_{j \in k}B_{ij} x_j^{(k)} &amp;=1\\
\left(B^{(k, k)} x^{(k)}\right)_i &amp;=1\\
B^{(k, k)} x^{(k)} &amp;= 1_{\|k\|}
\end{aligned}
\]</span></p>
<p>But this is exactly the structure of the equilibria for a GLV model, which we have introduced before.</p>
</div>
<div id="the-structure-of-equilibria-in-glv" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> The structure of equilibria in GLV<a href="predicting-coexistence-in-ecological-communities.html#the-structure-of-equilibria-in-glv" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Take the model:</p>
<p><span class="math display">\[
\dot{x} = D(x \circ r)(1_n - Bx)
\]</span></p>
<p>with <span class="math inline">\(r &gt; 0_n\)</span>; compute all the feasible equilibria and collect them into a matrix <span class="math inline">\(E\)</span>. For example:</p>
<p><span class="math display">\[
B = \begin{pmatrix}
2 &amp; -3 &amp; 3\\
-2 &amp; 7 &amp; -3\\
2 &amp; -2 &amp; 2
\end{pmatrix} \quad
E = \begin{pmatrix}
\frac{1}{2} &amp; 0 &amp; 0\\
0 &amp; \frac{1}{7} &amp; 0\\
0 &amp; 0 &amp; \frac{1}{2}\\
\frac{5}{4} &amp;\frac{1}{2} &amp; 0\\
0 &amp; \frac{5}{8} &amp; \frac{9}{8}\\
\frac{1}{2} &amp; \frac{1}{2} &amp; \frac{1}{2}
\end{pmatrix}
\]</span></p>
<p>where each row contains a feasible equilibrium (if it exists) corresponding to a given composition.</p>
<div class="boxpaper">
<p><strong>Key paper: <span class="citation">Tilman et al. (<a href="#ref-tilman1997plant">1997</a>)</span></strong></p>
<p>We are not the first people to consider “in-silico” biodiversity experiments. Tilman and collaborators were doing the same more than twenty years ago! For a more sophisticated model, see <span class="citation">Loreau (<a href="#ref-loreau1998biodiversity">1998</a>)</span>.</p>
</div>
<p>Consider the matrix:</p>
<p><span class="math display">\[
1_m 1_n^T - E B^T = \begin{pmatrix}
\color{red} 0 &amp; 2 &amp; 0\\
\frac{10}{7} &amp; \color{red} 0 &amp; \frac{9}{7}\\
-\frac{1}{2} &amp;\frac{5}{2} &amp;\color{red}0\\
\color{red} 0 &amp; \color{red} 0 &amp; -\frac{1}{2}\\
-\frac{1}{2} &amp; \color{red} 0 &amp; \color{red} 0\\
\color{red} 0 &amp;\color{red} 0 &amp;\color{red} 0  
\end{pmatrix}
\]</span></p>
<p>For each row, we find <span class="math inline">\(0\)</span> for the corresponding population at equilibrium (in red), and the remaining values express <span class="math inline">\((1_n - B x^{(k)})_j\)</span>, which has the same sign as the invasion growth rates for population <span class="math inline">\(j\)</span>, when <span class="math inline">\(r &gt; 0_n\)</span>. Hence, a saturated equilibrium will correspond to a row with non positive values (e.g., the community <span class="math inline">\(k=\{1,2\}\)</span> is saturated, because population 3 cannot invade when rare).</p>
<p>Next, call <span class="math inline">\(E_i\)</span> the sub matrix obtained selecting only rows for which <span class="math inline">\(x_i^{(k)} &gt; 0\)</span>. For example, for population 1:</p>
<p><span class="math display">\[
E_1 = \begin{pmatrix}
\frac{1}{2} &amp; 0 &amp; 0\\
\frac{5}{4} &amp;\frac{1}{2} &amp; 0\\
\frac{1}{2} &amp; \frac{1}{2} &amp; \frac{1}{2}
\end{pmatrix}
\]</span>
We have that:</p>
<p><span class="math display">\[
E_i B_i = 1_l
\]</span></p>
<p>Where <span class="math inline">\(B_i\)</span> is the <span class="math inline">\(i^\text{th}\)</span> row of <span class="math inline">\(B\)</span>, and <span class="math inline">\(l\)</span> is the number of experiments in which <span class="math inline">\(i\)</span> is present:</p>
<p><span class="math display">\[
E_1 B_1 = \begin{pmatrix}
\frac{1}{2} &amp; 0 &amp; 0\\
\frac{5}{4} &amp;\frac{1}{2} &amp; 0\\
\frac{1}{2} &amp; \frac{1}{2} &amp; \frac{1}{2}
\end{pmatrix} \begin{pmatrix}
2\\
-3\\
3
\end{pmatrix} = \begin{pmatrix}
1\\
1\\
1
\end{pmatrix}
\]</span></p>
<p>Practically, this means that all the <span class="math inline">\(i^\text{th}\)</span> component of all the equilibria <span class="math inline">\(x^{(k)}\)</span> of a GLV model belong to a <em>hyperplane</em> in <span class="math inline">\(\mathbb R^n\)</span> defined by <span class="math inline">\(\sum_{j} B_{ij} x^{(k)}_j = 1\)</span>.</p>
<p>This also means that we can recover the values of <span class="math inline">\(B_i\)</span> from the equilibria:</p>
<p><span class="math display">\[
B_i = E_i^{-1}1_n
\]</span></p>
<p>which can be computed when <span class="math inline">\(E_i\)</span> is square.</p>
<div class="boxmath">
<p><strong>Moore-Penrose pseudoinverses</strong></p>
<p>For a <span class="math inline">\(n \times n\)</span> matrix <span class="math inline">\(A\)</span>, the inverse is determined (when it exists), as a matrix <span class="math inline">\(A^{-1}\)</span> such that <span class="math inline">\(AA^{-1} = A^{-1}A = I\)</span>. A square matrix is invertible if its determinant (remember, the product of its eigenvalues) is nonzero. A matrix with determinant zero is called <em>singular</em> or <em>degenerate</em> and has no inverse. Singular matrices are rare, in the sense that a random matrix is “almost never” singular (this can be stated in a precise mathematical way). An invertible matrix has <em>full rank</em> (i.e., the rows [columns] are all linearly independent).</p>
<p>Can we find a matrix that “works like” an inverse when <span class="math inline">\(A\)</span> is singular (or not square)? Turns out, we can find a matrix <span class="math inline">\(A^+\)</span> that satisfies these four criteria:</p>
<p><span class="math display">\[
AA^+A = A
\]</span>
<span class="math display">\[
A^+AA^+ = A^+
\]</span></p>
<p><span class="math display">\[AA^+ = (AA^+)^T\]</span>
<span class="math display">\[A^+A = (A^+A)^T\]</span></p>
<p>Suppose that <span class="math inline">\(A\)</span> has linearly independent columns (i.e., it is of full column rank), then <span class="math inline">\(A^T A\)</span> has full rank and is therefore invertible. In this case, we can compute the <em>left</em> pseudo-inverse as</p>
<p><span class="math display">\[A^+ = (A^T A)^{-1}A^T\]</span></p>
<p>This is called the left inverse because <span class="math inline">\(A^+A= (A^T A)^{-1} A^T A=I\)</span>. When <span class="math inline">\(A\)</span> has independent rows, one can compute the <em>right inverse</em>. Note that a generalized inverse <span class="math inline">\(A^+\)</span> exists even if <span class="math inline">\(A\)</span> is not of full column nor full row rank, but in this case it is not unique.</p>
<p><strong>Application: least squares</strong></p>
<p>The pseudo-inverse can be used to find the <em>least-squares solution</em> of a system of linear equations. For example, in linear regression, we want to model a set of <span class="math inline">\(n\)</span> observations, <span class="math inline">\(y\)</span>, as a linear function of measured predictors <span class="math inline">\(X\)</span> (for example, a matrix with <span class="math inline">\(n\)</span> rows [one for each observation], and <span class="math inline">\(k\)</span> columns [the number of measured predictors for each observation, typically with <span class="math inline">\(k \ll n\)</span>]) and some parameters <span class="math inline">\(b\)</span>. The linear system</p>
<p><span class="math display">\[
Xb = y
\]</span></p>
<p>has no solution (for example, because <span class="math inline">\(X\)</span> is rectangular). If <span class="math inline">\(X\)</span> were to be invertible, finding the parameters would be easy:</p>
<p><span class="math display">\[
\hat{b} = X^{-1}y
\]</span></p>
<p>we can attempt the same approach using the pseudo-inverse:</p>
<p><span class="math display">\[
\hat{b} = X^{+}y
\]</span></p>
<p>in particular, it can be proven that the solution <span class="math inline">\(\hat{b}\)</span> minimizes the sum of squared deviations (SSQ): call <span class="math inline">\(\hat{y} = X \hat{b}\)</span>, then the solution <span class="math inline">\(\hat{b}\)</span> is the parameter choice minimizing <span class="math inline">\(SSQ = \sum_i (y_i -\hat{y}_i)^2\)</span>.</p>
</div>
<p>If <span class="math inline">\(E_i\)</span> is rectangular, we can use the Moore-Penrose pseudo-inverse. For example:</p>
<p><span class="math display">\[
E_2 = \begin{pmatrix}
0 &amp; \frac{1}{7} &amp; 0\\
\frac{5}{4} &amp;\frac{1}{2} &amp; 0\\
0 &amp; \frac{5}{8} &amp; \frac{9}{8}\\
\frac{1}{2} &amp; \frac{1}{2} &amp; \frac{1}{2}
\end{pmatrix}
\]</span></p>
<p><span class="math display">\[
(E_2^T E_2)^{-1}E_2^T 1_4= \begin{pmatrix}
-2 \\
7 \\
-3
\end{pmatrix} = B_2
\]</span></p>
<p>This also provides a naive method to get a draft of <span class="math inline">\(B\)</span> when only the equilibrium values are available.</p>
<p>The method is going to return a unique solution provided that each matrix <span class="math inline">\(E_i\)</span> has rank <span class="math inline">\(n\)</span>. In practice, this means that, for each population <span class="math inline">\(i\)</span>, a) <span class="math inline">\(E_i\)</span> has at least <span class="math inline">\(n\)</span> rows (i.e., population <span class="math inline">\(i\)</span> is present at <span class="math inline">\(n\)</span> distinct equilibria), and b) each species <span class="math inline">\(j\)</span> is present in at least one equilibrium with <span class="math inline">\(i\)</span>, i.e., each pair of populations is found in at least one equilibrium.</p>
<p>This might be a tall order when <span class="math inline">\(n\)</span> is large, and we will see how to circumvent this issue below.</p>
</div>
<div id="predicting-coexistence" class="section level2 hasAnchor" number="5.5">
<h2><span class="header-section-number">5.5</span> Predicting coexistence<a href="predicting-coexistence-in-ecological-communities.html#predicting-coexistence" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Suppose that we have estimated the matrix <span class="math inline">\(B\)</span>, and that we want to determine whether a certain combination of populations can or cannot coexist. This is useful when we have sufficient data to estimate <span class="math inline">\(B\)</span>, but we have not carried out all the possible experiments—we want to predict the outcome of further experiments in advance.</p>
<p>Then, according to our statistical model, we can take <span class="math inline">\(B^{(k,k)}\)</span> for the desired community <span class="math inline">\(k\)</span>, and take the row sums of the inverse:</p>
<p><span class="math display">\[
x^{(k)} = \left( B^{(k,k)} \right)^{-1} 1_{\|k\|}
\]</span></p>
<p>If all the components of <span class="math inline">\(x^{(k)}\)</span> are positive, we assume that this will be the outcome of the experiment. If <em>any</em> of the components is negative, on the other hand, we conclude that the populations cannot coexist.</p>
<div class="boxhomework">
<p><strong>Homework 6a</strong></p>
<p>As seen above, for GLV, the matrix <span class="math inline">\(B = D(r)^{-1}A\)</span> encodes all the equilibria (feasible or unfeasible) for the model. Knowing <span class="math inline">\(B\)</span> is therefore all we need to determine the existence of a feasible equilibrium for any community we can form from the pool.</p>
<ol style="list-style-type: decimal">
<li><p>Does the matrix <span class="math inline">\(B\)</span> informs us on invasibility? Take a sub-community <span class="math inline">\(k\)</span> resting at its feasible equilibrium point. Can one determine whether a species <span class="math inline">\(i \notin k\)</span> can invade when rare by inspecting the matrix <span class="math inline">\(B\)</span>?</p></li>
<li><p>When can the matrix <span class="math inline">\(B\)</span> be used to determine the stability of the feasible equilibrium for a subset <span class="math inline">\(k\)</span>?</p></li>
</ol>
</div>
</div>
<div id="scoring-matrices-b" class="section level2 hasAnchor" number="5.6">
<h2><span class="header-section-number">5.6</span> Scoring matrices <span class="math inline">\(B\)</span><a href="predicting-coexistence-in-ecological-communities.html#scoring-matrices-b" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The paragraph above suggests a good way to score matrices in practice (i.e., when the empirically-observed matrix <span class="math inline">\(\tilde{E}\)</span> is a noisy estimate of a true matrix <span class="math inline">\(E\)</span>):</p>
<ul>
<li>Propose a matrix <span class="math inline">\(B\)</span></li>
<li>Compute <span class="math inline">\(x^{(k)}\)</span> for each observed <span class="math inline">\(\tilde{x}^{(k,r)}\)</span></li>
<li>Try to minimize the <em>distance</em> between the predicted and observed</li>
</ul>
<p>Naturally, this algorithm relies on a notion of distance. The simplest choice would be to choose <span class="math inline">\(B\)</span> such that the sum of squared deviations is minimized:</p>
<p><span class="math display">\[
SSQ(B) = \sum_r \sum_k \sum_i \left(\tilde{x}^{(k,r)}_i - x^{(k)}_i \right)^2
\]</span></p>
<p>When the values of <span class="math inline">\(\tilde{x}^{(k)}_i\)</span> vary considerably (as expected when populations interact), minimizing the SSQ will favor matrices that match closely the rows of <span class="math inline">\(\tilde{E}\)</span> containing large values, such that a 10% error on a small value of <span class="math inline">\(\tilde{x}^{(k)}_i\)</span> “counts” less than a 1% error for a large value.</p>
<p>If we want to put all rows on the same footing, we can implement a Weighted Least Squares scheme, in which</p>
<p><span class="math display">\[
WLS(B) = \sum_r \sum_k \sum_i \left(\frac{\tilde{x}^{(k,r)}_i - x^{(k)}_i}{\sigma^{(k)}_i} \right)^2
\]</span></p>
<p>where deviations are weighted by the respective standard deviation (for simplicity, we can take <span class="math inline">\(\sigma_i^{(k)} = \sqrt{ \mathbb E\left((\tilde{x}_i^{(k,r)})^2 \right) - \left( \mathbb E\left(\tilde{x}_i^{(k,r)} \right)\right)^2}\)</span>.</p>
<p>Finally, we can have a likelihood-based approach in which we are trying to maximize the sum of log-likelihoods:</p>
<p><span class="math display">\[
\mathcal L(B) = \sum_r \sum_k \sum_i \log P\left(\tilde{x}^{(k,r)}_i|x^{(k)}_i, \gamma_i^{(k)} \right)
\]</span></p>
<p>where <span class="math inline">\(P(x|\mu, \gamma)\)</span> is the density of a probability distribution function whose shape is controlled by parameters <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\gamma\)</span>.</p>
<div class="boxpaper">
<p><strong>Key paper: <span class="citation">Loreau and Hector (<a href="#ref-loreau2001partitioning">2001</a>)</span></strong></p>
<p>Ecologists are fond of the idea of partitioning effects into (hopefully) orthogonal components that have a clear biological interpretation. In this influential paper, Loreau and Hector show how the effect of diversity on functioning can be partitioned into “selection” (dominance of species with particular traits) vs. “complementarity” (increased resource use due to niche partitioning). As for any interesting idea, critics abound—and some are over the top (<span class="citation">Pillai and Gouhier (<a href="#ref-pillai2019not">2019a</a>)</span>), leading to rebuttals of rebuttals (<span class="citation">Loreau and Hector (<a href="#ref-loreau2019not">2019</a>)</span>, <span class="citation">Wagg et al. (<a href="#ref-wagg2019not">2019</a>)</span>, <span class="citation">Pillai and Gouhier (<a href="#ref-pillai2019not3">2019b</a>)</span>, <span class="citation">Pillai and Gouhier (<a href="#ref-pillai2019not2">2019c</a>)</span>)…</p>
</div>
</div>
<div id="example" class="section level2 hasAnchor" number="5.7">
<h2><span class="header-section-number">5.7</span> Example<a href="predicting-coexistence-in-ecological-communities.html#example" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We are going to use the code accompanying the paper</p>
<ul>
<li>Skwara, A., Lemos‐Costa, P., Miller, Z.R. and Allesina, S., 2023. <a href="https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.14028">Modelling ecological communities when composition is manipulated experimentally</a>. Methods in Ecology and Evolution, 14(2), pp.696-707.</li>
</ul>
<p>which you can find here</p>
<p><a href="https://github.com/StefanoAllesina/skwara_et_al_2022/"><code>https://github.com/StefanoAllesina/skwara_et_al_2022/</code></a></p>
<p>we have slightly massaged the code for this tutorial.</p>
<div id="sum-of-squares" class="section level3 hasAnchor" number="5.7.1">
<h3><span class="header-section-number">5.7.1</span> Sum of squares<a href="predicting-coexistence-in-ecological-communities.html#sum-of-squares" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, we are going to use all the data to find a matrix <span class="math inline">\(B\)</span> that best encodes the density of the populations. To this end, we try to minimize the SSQ</p>
<p><span class="math display">\[
SSQ(B) = \sum_r \sum_k \sum_i \left(\tilde{x}^{(k,r)}_i - x^{(k)}_i \right)^2
\]</span></p>
<p>This can be accomplished by calling:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="predicting-coexistence-in-ecological-communities.html#cb2-1" tabindex="-1"></a>full_SSQ <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb2-2"><a href="predicting-coexistence-in-ecological-communities.html#cb2-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb2-3"><a href="predicting-coexistence-in-ecological-communities.html#cb2-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;full&quot;</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb2-4"><a href="predicting-coexistence-in-ecological-communities.html#cb2-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;SSQ&quot;</span>, <span class="co"># minimize Sum of Squared Deviations</span></span>
<span id="cb2-5"><a href="predicting-coexistence-in-ecological-communities.html#cb2-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb2-6"><a href="predicting-coexistence-in-ecological-communities.html#cb2-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span> <span class="co"># go directly to numerical optimization</span></span>
<span id="cb2-7"><a href="predicting-coexistence-in-ecological-communities.html#cb2-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597</p>
<p>The values being printed are the SSQ after each round of numerical optimization (in this case, the calculation converges immediately to the solution).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="predicting-coexistence-in-ecological-communities.html#cb3-1" tabindex="-1"></a><span class="fu">plot_results_boxplot</span>(full_SSQ<span class="sc">$</span>observed, full_SSQ<span class="sc">$</span>predicted)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-36-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>The boxplots show the data (points), as well as the corresponding boxplots, with the horizontal line being the median value of <span class="math inline">\(\tilde{x}_i^{(k,r)}\)</span> across <span class="math inline">\(r\)</span>, the triangle shows the location of the empirical mean. The circle marks the fitted mean value for the combination of strain/community, obtained computing <span class="math inline">\(\left(B^{(k,k)}\right)^{-1}1_{\|k\|}\)</span>. As you can see, we can find a matrix <span class="math inline">\(B\)</span> that recapitulates the observed means quite well.</p>
<p>Let’s plot the predicted vs. observed values:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="predicting-coexistence-in-ecological-communities.html#cb4-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_SSQ)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-37-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Where the points mark the predicted vs observed means, and the crosses the full data. The dashed line is the 1:1 line.</p>
</div>
<div id="weighted-leas-squares" class="section level3 hasAnchor" number="5.7.2">
<h3><span class="header-section-number">5.7.2</span> Weighted leas squares<a href="predicting-coexistence-in-ecological-communities.html#weighted-leas-squares" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can repeat the calculation, but this time trying to minimize</p>
<p><span class="math display">\[
WLS(B) = \sum_r \sum_k \sum_i \left(\frac{\tilde{x}^{(k,r)}_i - x^{(k)}_i}{\sigma^{(k)}_i} \right)^2
\]</span></p>
<p>by calling:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="predicting-coexistence-in-ecological-communities.html#cb5-1" tabindex="-1"></a>full_WLS <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb5-2"><a href="predicting-coexistence-in-ecological-communities.html#cb5-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb5-3"><a href="predicting-coexistence-in-ecological-communities.html#cb5-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;full&quot;</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb5-4"><a href="predicting-coexistence-in-ecological-communities.html#cb5-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;WLS&quot;</span>, </span>
<span id="cb5-5"><a href="predicting-coexistence-in-ecological-communities.html#cb5-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb5-6"><a href="predicting-coexistence-in-ecological-communities.html#cb5-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb5-7"><a href="predicting-coexistence-in-ecological-communities.html#cb5-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833</p>
<p>Plotting:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="predicting-coexistence-in-ecological-communities.html#cb6-1" tabindex="-1"></a><span class="fu">plot_results_boxplot</span>(full_WLS<span class="sc">$</span>observed, full_WLS<span class="sc">$</span>predicted)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-39-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="predicting-coexistence-in-ecological-communities.html#cb7-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_WLS)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-39-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>Notice that the points have moved slightly—this is because we are penalizing deviations differently depending on the measured variance (and thus points with a higher variance can depart more strongly from the 1:1 line).</p>
</div>
<div id="maximum-likelihood" class="section level3 hasAnchor" number="5.7.3">
<h3><span class="header-section-number">5.7.3</span> Maximum likelihood<a href="predicting-coexistence-in-ecological-communities.html#maximum-likelihood" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now we take a different approach, and take the observations to be independent samples from a log-normal distribution:</p>
<p><span class="math display">\[
\tilde{x}_i^{(k,r)} \sim LN(x_i^{(k)}, \sigma_i)
\]</span></p>
<p>i.e., we take each variable to have a mean determined by the corresponding <span class="math inline">\(x_i^{(k)}\)</span>, and a variance parameter that depends only on strain identity.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="predicting-coexistence-in-ecological-communities.html#cb8-1" tabindex="-1"></a>full_LN <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb8-2"><a href="predicting-coexistence-in-ecological-communities.html#cb8-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb8-3"><a href="predicting-coexistence-in-ecological-communities.html#cb8-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;full&quot;</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb8-4"><a href="predicting-coexistence-in-ecological-communities.html#cb8-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;LikLN&quot;</span>, </span>
<span id="cb8-5"><a href="predicting-coexistence-in-ecological-communities.html#cb8-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb8-6"><a href="predicting-coexistence-in-ecological-communities.html#cb8-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span> <span class="co"># go directly to numerical optimization</span></span>
<span id="cb8-7"><a href="predicting-coexistence-in-ecological-communities.html#cb8-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="predicting-coexistence-in-ecological-communities.html#cb9-1" tabindex="-1"></a><span class="fu">plot_results_boxplot</span>(full_LN<span class="sc">$</span>observed, full_LN<span class="sc">$</span>predicted)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-41-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="predicting-coexistence-in-ecological-communities.html#cb10-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_LN)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-41-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>Also in this case, we obtain a good fit.</p>
</div>
<div id="leave-one-out-cross-validation" class="section level3 hasAnchor" number="5.7.4">
<h3><span class="header-section-number">5.7.4</span> Leave-one-out cross validation<a href="predicting-coexistence-in-ecological-communities.html#leave-one-out-cross-validation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The ultimate test for this type of model is to be able to predict experimental results before running the experiment.</p>
<p>In our case, we can try to leave out one of the 7 communities:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="predicting-coexistence-in-ecological-communities.html#cb11-1" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="predicting-coexistence-in-ecological-communities.html#cb12-1" tabindex="-1"></a>LOO_LN <span class="ot">&lt;-</span> <span class="fu">run_model_LOO</span>(<span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, </span>
<span id="cb12-2"><a href="predicting-coexistence-in-ecological-communities.html#cb12-2" tabindex="-1"></a>                        <span class="at">model =</span> <span class="st">&quot;full&quot;</span>, <span class="at">goalf =</span> <span class="st">&quot;LikLN&quot;</span>, </span>
<span id="cb12-3"><a href="predicting-coexistence-in-ecological-communities.html#cb12-3" tabindex="-1"></a>                        <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="at">plot_results =</span> <span class="cn">TRUE</span>, </span>
<span id="cb12-4"><a href="predicting-coexistence-in-ecological-communities.html#cb12-4" tabindex="-1"></a>                        <span class="at">LOO_row_num =</span>  <span class="dv">40</span> <span class="co"># exclude all experiments with this community</span></span>
<span id="cb12-5"><a href="predicting-coexistence-in-ecological-communities.html#cb12-5" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928</p>
<p>The model is quite successful at recovering the observed means for the community we left out.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="predicting-coexistence-in-ecological-communities.html#cb13-1" tabindex="-1"></a><span class="fu">plot_results_boxplot</span>(LOO_LN<span class="sc">$</span>observed, LOO_LN<span class="sc">$</span>predicted, LOO_LN<span class="sc">$</span>infit)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-44-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="simplified-models" class="section level3 hasAnchor" number="5.7.5">
<h3><span class="header-section-number">5.7.5</span> Simplified models<a href="predicting-coexistence-in-ecological-communities.html#simplified-models" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We have seen in the lecture that this approach estimates all <span class="math inline">\(n^2\)</span> coefficients of the matrix <span class="math inline">\(B\)</span>. For the approach to be successful, we need to have observed enough combinations of populations growing together (each population must be present in at least <span class="math inline">\(n\)</span> experiments with distinct compositions, each pair of populations must appear in at least one final community). When the number of populations in the pool is large, and experiments are few, this approach is unfeasible. We can therefore try to simplify the model to reduce the number of parameters.</p>
<p>The idea brought forward by Skwara <em>et al.</em> is to approximate the matrix <span class="math inline">\(B\)</span> as the sum of a diagonal matrix and a low-rank matrix. For example, a version of the model with only <span class="math inline">\(n + 1\)</span> parameters reads:</p>
<p><span class="math display">\[
B = D(s) + \alpha 1_n1_n^T
\]</span></p>
<p>i.e., a model in which each diagonal element <span class="math inline">\(B_{ii} = s_i + \alpha\)</span> has its own parameter (and thus can take arbitrary values), while the off-diagonals are all the same (i.e., each population has the same “mean-field” effect on all others).</p>
<p>A more general model using <span class="math inline">\(3 n - 1\)</span> parameters reads:</p>
<p><span class="math display">\[
B = D(s) + u v^T
\]</span></p>
<p>in which the diagonal coefficients <span class="math inline">\(B_{ii} = s_i + u_i v_i\)</span> can take arbitrary values, while the off-diagonal elements <span class="math inline">\(B_{ij} = u_i v_j\)</span> are constrained; the effect of species <span class="math inline">\(j\)</span> on <span class="math inline">\(i\)</span> depends on two values: <span class="math inline">\(v_i\)</span> that measures how strongly species <span class="math inline">\(i\)</span> respond to the presence of other species, and <span class="math inline">\(v_j\)</span> that measures the magnitude of the typical effect of species <span class="math inline">\(j\)</span>. Either <span class="math inline">\(u\)</span> or <span class="math inline">\(v\)</span> can be taken as unitary (i.e., <span class="math inline">\(\sum_i v_i^2 = 1\)</span>) without loss of generality, thereby bringing the total number of parameters to <span class="math inline">\(3n-1\)</span>. When the number of populations is small, this approach does not lead to big gains (e.g., for <span class="math inline">\(n=3\)</span> we have 8 parameters instead of 9), but the reduction in parameters is substantial when <span class="math inline">\(n\)</span> becomes larger (e.g., for the whole data set, <span class="math inline">\(n=7\)</span>, and thus we would use 20 parameters instead of 49).</p>
<p>Another advantage of this approach is that we can write the inverse in linear time, thereby removing the only big computational hurdle of the approach:</p>
<p><span class="math display">\[\begin{aligned}
B^{-1} 1_n&amp;= (D(s) + u v^T)^{-1}1_n\\
&amp;= \left(D(s)^{-1} - \frac{1}{1 + v^TD(s)^{-1}u}D(s)^{-1} u v^TD(s)^{-1}\right)1_n\\ &amp;= D(s)^{-1}\left(1_n - \frac{v^T D(s)^{-1}1_n}{1 + v^TD(s)^{-1}u} u\right)
\end{aligned}
\]</span></p>
<p>These models can be derived from the consumer-resource framework that you have seen in class.</p>
<p>We can test different versions of the simplified model. A model with only the diagonal and an extra parameter performs very poorly:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="predicting-coexistence-in-ecological-communities.html#cb14-1" tabindex="-1"></a>diag_a11 <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb14-2"><a href="predicting-coexistence-in-ecological-communities.html#cb14-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb14-3"><a href="predicting-coexistence-in-ecological-communities.html#cb14-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;diag_a11t&quot;</span>,</span>
<span id="cb14-4"><a href="predicting-coexistence-in-ecological-communities.html#cb14-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;LikLN&quot;</span>, </span>
<span id="cb14-5"><a href="predicting-coexistence-in-ecological-communities.html#cb14-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb14-6"><a href="predicting-coexistence-in-ecological-communities.html#cb14-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span></span>
<span id="cb14-7"><a href="predicting-coexistence-in-ecological-communities.html#cb14-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="predicting-coexistence-in-ecological-communities.html#cb15-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_a11)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-46-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>A model with symmetric interactions, <span class="math inline">\(B = D(s) + vv^T\)</span> does slightly better:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="predicting-coexistence-in-ecological-communities.html#cb16-1" tabindex="-1"></a>diag_vvt <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb16-2"><a href="predicting-coexistence-in-ecological-communities.html#cb16-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb16-3"><a href="predicting-coexistence-in-ecological-communities.html#cb16-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;diag_vvt&quot;</span>,</span>
<span id="cb16-4"><a href="predicting-coexistence-in-ecological-communities.html#cb16-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;LikLN&quot;</span>, </span>
<span id="cb16-5"><a href="predicting-coexistence-in-ecological-communities.html#cb16-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb16-6"><a href="predicting-coexistence-in-ecological-communities.html#cb16-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span></span>
<span id="cb16-7"><a href="predicting-coexistence-in-ecological-communities.html#cb16-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] -38.12933
[1] -38.36503
[1] -38.37226
[1] -38.38069
[1] -38.38508
[1] -38.39331
[1] -38.39618
[1] -38.39655
[1] -38.39655
[1] -38.39655</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="predicting-coexistence-in-ecological-communities.html#cb17-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_vvt)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-48-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>And, finally, the model with 8 parameters, <span class="math inline">\(B = D(s) + vw^T\)</span>:</p>
<p>A model with symmetric interactions, <span class="math inline">\(B = D(s) + vv^T\)</span> does slightly better:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="predicting-coexistence-in-ecological-communities.html#cb18-1" tabindex="-1"></a>diag_vwt <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb18-2"><a href="predicting-coexistence-in-ecological-communities.html#cb18-2" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">&quot;dat/Ishizawa_3_strains.csv&quot;</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb18-3"><a href="predicting-coexistence-in-ecological-communities.html#cb18-3" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;diag_vwt&quot;</span>,</span>
<span id="cb18-4"><a href="predicting-coexistence-in-ecological-communities.html#cb18-4" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">&quot;LikLN&quot;</span>, </span>
<span id="cb18-5"><a href="predicting-coexistence-in-ecological-communities.html#cb18-5" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb18-6"><a href="predicting-coexistence-in-ecological-communities.html#cb18-6" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span></span>
<span id="cb18-7"><a href="predicting-coexistence-in-ecological-communities.html#cb18-7" tabindex="-1"></a>)</span></code></pre></div>
<p>[1] “numerical search”
[1] -34.87107
[1] -49.06619
[1] -49.06753
[1] -49.0681
[1] -49.0681
[1] -49.0681
[1] -49.0681
[1] -49.06811
[1] -49.06811
[1] -49.06811</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="predicting-coexistence-in-ecological-communities.html#cb19-1" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_vwt)</span></code></pre></div>
<p><img src="Theoretical_Community_Ecology_files/figure-html/unnamed-chunk-50-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="further-readings-2" class="section level2 hasAnchor" number="5.8">
<h2><span class="header-section-number">5.8</span> Further readings<a href="predicting-coexistence-in-ecological-communities.html#further-readings-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The backbone of the statistical model can be found in numerous articles, which converged to the same solution starting from different angles:</p>
<ul>
<li><p>Xiao, Y., Angulo, M. T., Friedman, J., Waldor, M. K., Weiss, S. T., &amp; Liu, Y.-Y. (2017). <a href="https://www.nature.com/articles/s41467-017-02090-2">Mapping the ecological networks of microbial communities</a>. Nature Communications, 8(1), 1–12.</p></li>
<li><p>Fort, H. (2018). <a href="https://www.sciencedirect.com/science/article/abs/pii/S0304380018303053">On predicting species yields in multispecies communities: Quantifying the accuracy of the linear Lotka-Volterra generalized model</a>. Ecological Modelling, 387, 154–162.</p></li>
<li><p>Maynard, D. S., Miller, Z. R., &amp; Allesina, S. (2020). <a href="https://www.nature.com/articles/s41559-019-1059-z">Predicting coexistence in experimental ecological communities</a>. Nature Ecology &amp; Evolution, 4(1), 91–100.</p></li>
<li><p>Ansari, A. F., Reddy, Y., Raut, J., &amp; Dixit, N. M. (2021). <a href="https://www.nature.com/articles/s43588-021-00131-x">An efficient and scalable top-down method for predicting structures of microbial communities</a>. Nature Computational Science, 1(9), 619–628.</p></li>
<li><p>Davis, J.D., Olivença, D.V., Brown, S.P. and Voit, E.O., (2022). <a href="https://www.frontiersin.org/journals/systems-biology/articles/10.3389/fsysb.2022.1021897/full">Methods of quantifying interactions among populations using Lotka-Volterra models</a>. Frontiers in Systems Biology, 2, p.1021897.</p></li>
</ul>
<p>For the tutorial, we are going to closely follow:</p>
<ul>
<li>Skwara, A., Lemos‐Costa, P., Miller, Z.R. and Allesina, S., 2023. <a href="https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.14028">Modelling ecological communities when composition is manipulated experimentally</a>. Methods in Ecology and Evolution, 14(2), pp.696-707.</li>
</ul>
<p>The same method can be extended to provide a simple test for phylogenetic effects on competition/growth:</p>
<ul>
<li>Lemos‐Costa, P., Miller, Z.R. and Allesina, S., 2024. <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/ele.14490">Phylogeny structures species’ interactions in experimental ecological communities</a>. Ecology Letters, 27(8), p.e14490.</li>
</ul>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-cadotte2013experimental" class="csl-entry">
Cadotte, M. W. 2013. Experimental evidence that evolutionarily diverse assemblages result in higher productivity. Proceedings of the National Academy of Sciences 110:8996–9000.
</div>
<div id="ref-hector1999plant" class="csl-entry">
Hector, A., B. Schmid, C. Beierkuhnlein, M. Caldeira, M. Diemer, P. Dimitrakopoulos, J. Finn, H. Freitas, P. Giller, J. Good, and others. 1999. Plant diversity and productivity experiments in european grasslands. science 286:1123–1127.
</div>
<div id="ref-loreau1998biodiversity" class="csl-entry">
Loreau, M. 1998. Biodiversity and ecosystem functioning: A mechanistic model. Proceedings of the National Academy of Sciences 95:5632–5636.
</div>
<div id="ref-loreau2001partitioning" class="csl-entry">
Loreau, M., and A. Hector. 2001. Partitioning selection and complementarity in biodiversity experiments. Nature 412:72–76.
</div>
<div id="ref-loreau2019not" class="csl-entry">
Loreau, M., and A. Hector. 2019. Not even wrong: Comment by loreau and hector. Ecology 100:e02794.
</div>
<div id="ref-pillai2019not3" class="csl-entry">
Pillai, P., and T. C. Gouhier. 2019b. Not even wrong: Reply to loreau and hector. arXiv preprint arXiv:1910.13563.
</div>
<div id="ref-pillai2019not2" class="csl-entry">
Pillai, P., and T. C. Gouhier. 2019c. Not even wrong: Reply to wagg et al. arXiv preprint arXiv:1910.13670.
</div>
<div id="ref-pillai2019not" class="csl-entry">
Pillai, P., and T. C. Gouhier. 2019a. Not even wrong: The spurious measurement of biodiversity’s effects on ecosystem functioning. Ecology 100:e02645.
</div>
<div id="ref-tilman1997plant" class="csl-entry">
Tilman, D., C. L. Lehman, and K. T. Thomson. 1997. Plant diversity and ecosystem productivity: Theoretical considerations. Proceedings of the national academy of sciences 94:1857–1861.
</div>
<div id="ref-tilman2001diversity" class="csl-entry">
Tilman, D., P. B. Reich, J. Knops, D. Wedin, T. Mielke, and C. Lehman. 2001. Diversity and productivity in a long-term grassland experiment. Science 294:843–845.
</div>
<div id="ref-wagg2019not" class="csl-entry">
Wagg, C., K. E. Barry, M. J. O’Brien, A. McKenzie-Gopsill, C. Roscher, N. Eisenhauer, and B. Schmid. 2019. Not even wrong: Comment by wagg et al. Ecology 100:e02805.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="stability-of-large-ecological-communities.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="game-theory-and-replicator-dynamics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Theoretical_Community_Ecology.pdf", "Theoretical_Community_Ecology.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
